\documentclass[letterpaper]{article}

% ##- In R -
% ## Sweave("bio3d_vignette.Rnw")
% ## tools::texi2dvi("bio3d_vignette.tex", pdf=TRUE)
% ##

\usepackage{natbib}
\usepackage{color}
\definecolor{myurlblue}{rgb}{0.3,0.2,0.7}
\usepackage[colorlinks=true,urlcolor=myurlblue,pagecolor=black,citecolor=myurlblue,linkcolor=black]{hyperref}


\title{An Introduction to the bio3d Package}
\author{Barry Grant\\
University of California, San Diego}



\begin{document}
\SweaveOpts{prefix.string=figs/fig}

\maketitle
\tableofcontents


\section{Introduction}
The aim of this document, termed a vignette\footnote{This vignette contains executable examples, see \texttt{help(vignette)} for further details.} in R parlance, is to provide a task-oriented introduction to the bio3d R package \citep{grant06}.  Bio3d contains utilities to process, organize and explore protein structure, sequence and dynamics data.  Features include the ability to read and write structure, sequence and dynamic trajectory data, perform atom selection, re-orientation, superposition, rigid core identification, clustering, torsion analysis, distance matrix analysis, structure and sequence conservation analysis, and principal component analysis.

In addition, various utility functions are provided to enable the statistical and graphical power of the R environment to work with biological sequence and structural data.  




\paragraph{Supporting Material:}
The latest version and further documentation can be obtained from the bio3d website: \href{http://mccammon.ucsd.edu/~bgrant/bio3d/}{http://mccammon.ucsd.edu/$\sim$bgrant/bio3d/} and wiki: \href{http://bio3d.pbwiki.com/}{http://bio3d.pbwiki.com/}.




\subsection{Getting Started}
Start R, load the bio3d package and use the command \texttt{lbio3d()} to list the current functions available within the package:

<<>>=
library(bio3d)
lbio3d()
@
Detailed documentation and example code for each function can be accessed via the \texttt{help()} and \texttt{example()} commands (e.g. \texttt{help(read.pdb)}).  You can also copy and paste any of the example code from the documentation of a particular function directly into your R session:

<<>>=
pdb <- read.pdb("1bg2")
@ 
In the above example the function  \texttt{read.pdb()} is supplied with one input argument, the four letter RCSB Protein Data Bank identifier \texttt{1bg2}. As this   
In the above example the \texttt{get.pdb()} command returns the RCSB Protein Data Bank file for structure \texttt{1bg2}.  Setting \texttt{URLonly} to \texttt{FALSE} would cause the file to be downloaded to the current working directory.  The \texttt{read.pdb()} command then processes this file and returns its output to the object \texttt{pdb}.  To examine the contents of the \texttt{pdb} object we can use the \texttt{attributes} command.

<<>>=
attributes(pdb)
@
Most bio3d functions, including \texttt{read.pdb()}, return list objects whose components can be accessed in the standard way (see \texttt{help(read.pdb)} for further details).  For example, to print coordinate data for the first three atoms in our newly created \texttt{pdb} object type:

<<>>=
pdb$atom[1:3, c("resno","resid","elety","x","y","z")]
@
In the previous example we use numeric indices to access atoms 1 to 3. In a similar fashion the \texttt{atom.select()} function returns numeric indices that can be used for convinent data access: 
<<fig=TRUE>>=
inds <- atom.select(pdb, "calpha")
plot.bio3d(pdb$atom[inds$atom, "resno"], pdb$atom[inds$atom, "b"], sse=pdb, ylab="Bfactor", xlab="Residue Number")
@
The returned \texttt{inds} object is a list containing atom and xyz numeric indices (in this case corresponding to all alpha Carbon atoms). In the above example we use these indices to plot residue number vs B-factor along with a basic secondary structure schematic. We will return to discuss atom selection in more detail in a subsequent section but as a further example of data access lets extract the sequence for the loop region between strand four and helix four in our \texttt{pdb} object.
<<>>=
loop <- pdb$sheet$end[4]:pdb$helix$start[4]
loop.inds <- atom.select(pdb,"///loop///CA/")
pdb$atom[loop.inds$atom,"resid"]
aa321(pdb$atom[loop.inds$atom,"resid"])
@ 
In the above example the residue numbers in the \texttt{sheet} and \texttt{helix} components of \texttt{pdb} are accessed and used in a subsequent atom selection. The \texttt{aa321()} function converts between three-letter and one-letter IUPAC aminoacid codes.

Note that \texttt{pdb} object can be passed to other bio3d functions without the need for atom selection, including \texttt{torsion.pdb()}, \texttt{convert.pdb()}, \texttt{dssp()} and the distance matrix function \texttt{dm()}:
% <<fig=TRUE>>=
% d <- dm(pdb)
% plot(d, xlab="Residue Index", ylab="Residue Index")
% @

\subsection{Example Data}
A number of example datasets are included with the bio3d package. The main purpose of including this data (which may be generated by the user by following the extended examples documented within the various bio3d functions) is to allow users to more quickly appreciate the capabilities of functions that would otherwise require data input and processing before execution.


\subsubsection{The Kinesin Molecular Motor}
For the worked examples in the current document we will utilize the included kinesin dataset.  A related dataset formed the basis of the work described in \citep{grant07}.  Briefly, kinesins are molecular motor proteins responsible for the ATP dependent transport of cellular cargo along microtubules.  Kinesin family members have been found in all eukaryotic organisms, where they contribute to the transport of molecules and organelles, organisation and maintenance of the cytoskeleton, and the segregation of genetic material during mitosis and meiosis. 

The defining attribute of kinesin family members is the possession of one or more globular motor domains. These $\sim$350 residue domains are responsible for ATP hydrolysis, microtubule binding and force production. The current dataset consists of kinesin motor domain sequence and structural data and can be loaded with the command \texttt{data(kinesin)}:

<<kinesin>>=
data(kinesin)
attach(kinesin)
@ 
\paragraph{Note:}The kinesin dataset can be assembled with the \texttt{read.fasta()} and \texttt{read.fasta.pdb()} commands bellow, which read a kinesin alignment and corresponding PDB structure files:

<<loadkinesin, eval=FALSE>>=
aln.file <- system.file("examples/kinesin_xray.fa",package="bio3d")
pdb.path <- system.file("examples/",package="bio3d")
aln  <- read.fasta(aln.file)
pdbs <- read.fasta.pdb(aln, pdb.path = pdb.path, pdbext = ".ent")
@



%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% --------------
% Second section 
% --------------

\section{Sample Applications of bio3d}
Comparing multiple structures of homologous proteins and carefully analysing large multiple sequence alignments can help identify patterns of sequence and structural conservation and highlight conserved interactions that are crucial for protein stability and function \citep{grant07}.  Bio3d and R provide a useful framework for such sudies and facilate the integration of sequence, structure and dynamics data in the analysis of protein evolution.


\subsection{Comparative Structural Analysis}
The detailed comparison of homologous protein structures can be used to infer pathways for evolutionary adaptation and, at closer evolutionary distances, mechanisms for conformational change. The bio3d package employs refined structural superposition and principal component analysis (PCA) to examine the relationship between different conformers. 

\subsubsection{Interconformer Relationships with PCA}
Conventional structural superposition of proteins minimizes the root mean square difference between their full set of equivalent residues. However, for the current application such a superposition procedure can be inappropriate. For example, in the comparison of a multi-domain protein that has undergone a hinge-like rearrangement of its domains, standard all atom superposition would result in an underestimate of the true atomic displacement by attempting superposition over all domains (whole structure superposition). A more appropriate and insightful superposition would be anchored at the most invariant region and hence more clearly highlight the domain rearrangement (sub-structure superposition). 


\paragraph{Sub-structure Superposition:}
The \texttt{core.find()} function implements an iterated superposition procedure, where residues displaying the largest positional differences are identified and excluded at each round.  The function returns an ordered list of excluded residues, from which the user can select a subset of 'core' residues upon which superposition can be based.

<<eval=FALSE>>=
core <- core.find(pdbs)
@ 
There are \texttt{plot.core()} and \texttt{print.core()} functions available for further examining the output of \texttt{core.find()}.
\begin{center}
<<fig=TRUE>>=
col=rep("black", length(core$volume))
col[core$volume<2]="pink"; col[core$volume<1]="red"
plot(core, col=col)
@
% \caption{Figure 1. core.find plot.}
\end{center}
The \texttt{print.core()} function also returns \texttt{atom} and \texttt{xyz} indices.  Below we select a subset of positions with a cumulative ellipsoid volume of less than 0.5 \AA $^3$ and use the returned indices to write a quick PDB file for viewing in a molecular graphics program (Figure 2).  
<<>>=
inds <- print(core, vol=0.5)
write.pdb(xyz=pdbs$xyz[1,inds$xyz], resno=pdbs$resno[1,inds$atom], file="quick_core.pdb")
@ 

\begin{center}
\includegraphics[width=60mm]{figs/core.png}
\end{center}
We can now superpose all structures on the selected core indices with the \texttt{fit.xyz()} function.
<<eval=FALSE>>=
xyz <- fit.xyz( fixed = pdbs$xyz[1,],
               mobile = pdbs,
               fixed.inds  = inds$xyz,
               mobile.inds = inds$xyz)
@
The above command performs the actual superposition and stores the new coordinats in the matrix object \texttt{xyz}.  By providing several extra arguments to \texttt{fit.xyz()} a directory, here named \texttt{fitlsq}, containing superposed structures can be produced.
<<>>=
xyz <- fit.xyz( fixed = pdbs$xyz[1,],
               mobile = pdbs,
               fixed.inds  = inds$xyz,
               mobile.inds = inds$xyz,
               pdb.path = system.file("examples/",package="bio3d"),
               pdbext = ".ent",
               outpath = "fitlsq/",
               full.pdbs = TRUE)
@ 
These can then be viewed in your favoriate molecular graphics program (Figure 3).

\begin{center}
\includegraphics[width=120mm]{figs/fit.png}
\end{center}

\paragraph{Principal Component Analysis:}
Following core identification and subsequent superposition, PCA can be employed to examine the relationship between different structures based on their equivalent residues. The application of PCA to both distributions of experimental structures and molecular dynamics trajectories, along with its ability to provide considerable insight into the nature of conformational differences has been discussed previously (see \citet{grant07} and references therein).  

Briefly, the resulting principal components (orthogonal eigenvectors) describe the axes of maximal variance of the distribution of structures. Projection of the distribution onto the subspace defined by the largest principal components results in a lower dimensional representation of the structural dataset. The percentage of the total mean square displacement (or variance) of atom positional fluctuations captured in each dimension is characterized by their corresponding eigenvalue. Experience suggests that 3--5 dimensions are often sufficient to capture over 70 percent of the total variance in a given family of structures. Thus, a handful of principal components are sufficient to provide a useful description while still retaining most of the variance in the original distribution \citet{grant06}. 

<<echo=FALSE>>=
cut.seqs <- which(pdbs$id %in% c("d1n6mb_","d1ry6a_"))
gaps.res <- gap.inspect(pdbs$ali[-cut.seqs,])
gaps.pos <- gap.inspect(pdbs$xyz[-cut.seqs,])
@ 

<<eval=FALSE>>=
## Remove outlier structures from kinesin alignment
cut.seqs <- which(pdbs$id %in% c("d1n6mb_","d1ry6a_"))

## Ignore gap containing positions
gaps.res <- gap.inspect(pdbs$ali[-cut.seqs,])
gaps.pos <- gap.inspect(pdbs$xyz[-cut.seqs,])

##-- Do PCA
pc.xray <- pca.xyz(xyz[-cut.seqs, gaps.pos$f.inds])
@ 
The above sequence of commands returns the indices of two structures and the indices for gap containing positions, both of which we exclude from subsequent PCA with the \texttt{pca.xyz()} command. A quick overview of the results of \texttt{pca.xyz()} can be obtained by calling \texttt{plot.pca()}

\begin{center}
<<fig=TRUE,echo=TRUE>>=
plot(pc.xray)
@ 
\end{center}

and calling  \texttt{plot.bio3d()} to examine the contribution of each residue to the first three principal components.

\begin{center}
<<fig=TRUE,echo=TRUE>>=
res.ref <- which(!is.gap(pdbs$ali["d1bg2__",]))
res.ind <- which(res.ref %in% gaps.res$f.ind)
par(mfrow = c(3, 1), cex = 0.6, mar = c(3, 4, 1, 1))
plot.bio3d(res.ind, pc.xray$au[,1], sse=sse, ylab="PC1 (A)")
plot.bio3d(res.ind, pc.xray$au[,2], sse=sse, ylab="PC2 (A)")
plot.bio3d(res.ind, pc.xray$au[,3], sse=sse, ylab="PC3 (A)")
@ 
\end{center}

The plots in figures 4 and 5 display the relationships between different conformers, highlight positions responsible for the major differences between structures and enable the interpretation and characterization of multiple interconformer relationships.


\paragraph{Structure Interpolation:}
To further aid interpretation, a PDB format trajectory can be produced that interpolates between the most dissimilar structures in the distribution along a given principal cmponent.  This involves dividing the difference between the conformers into a number of evenly spaced steps along the principal components, forming the frames of the trajectory. Such trajectories can be directly visualized in a molecular graphics program, such as VMD \citep{vmd}. Furthermore, the interpolated structures can be analyzed for possible domain and shear movements with the DynDom package \citep{dyndom}, or used as initial seed structures for more advanced reaction path refinement methods such as Conjugate Peak Refinement \citep{cpr}.

<<eval=FALSE>>=
a <- mktrj.pca(pc.xray, pc=1, file="pc1.pdb",
               resno = pdbs$resno[1, gaps.res$f.inds],
               resid = aa123(pdbs$ali[1, gaps.res$f.inds]) )
@ 

\begin{center}
\includegraphics[width=100mm]{figs/pca.png}
\end{center}

\paragraph{Conformer Clustering in PC space}
Clustering of structures in the PC1-3 planes

\begin{center}
<<fig=TRUE,echo=TRUE>>=
plot(hclust(dist(pc.xray$z[,1:3])) , labels=substr(pdbs$id[-cut.seqs ],2,10), main="PC1-3")
@ 
\end{center}




%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Third section %

\subsubsection{Further examples}
Bio3d facilitates the analysis of various structural properties, such as root mean-square deviation (RMSD), root mean-square fluctuation (RMSF), secondary structure, dihedral angles, difference distance matrices etc. The current section provides a brief exposure to using bio3d in this vein. 

\subsection{Finding available sets of similar structures}
First lets examine one way in which we can use bio3d to find sets of similar structures. To collect available kinesin crystal structures we first use BLAST to query the PDB to find similar sequences (and hence structures) to our chosen representative (PDB c ode 1BG2).

<<>>=
pdb <- read.pdb( get.pdb("4q21", URLonly=TRUE) )
seq <- seq.pdb(pdb)
blast <- blast.pdb(seq)
@ 
Examining the alignment scores and their associated E-values (with the function \texttt{plot.blast()}) indicates a sensible normalized score (-log(E-Value)) cutoff of ~150 bits.

\begin{figure}[hbtp]
\begin{center}
<<blasthits, fig=TRUE>>=
hits <- plot.blast(blast, cutoff=150)
head(hits)
@ 
\label{fig:blast}
\end{center}
\end{figure}


\paragraph{Multiple sequence alignment}
After downloading our complete list of structures from the PDB and splitting these into separate chains (see below) we extract the ATOM record sequence from each structure and align these to determine residue-residue correspondences. 

<<getpdbs, eval=FALSE>>=
unq.ids <- unique( substr(hits[,"pdb.id"], 1, 4) )
##- Download and chain split PDBs
raw.files <- get.pdb(unq.ids, path="raw_pdbs")
split.pdb(raw.files, path="raw_pdbs/split_chain/")
@ 


<<seqaln, eval=FALSE>>=
##- Extract and align sequences
files <- paste( "raw_pdbs/split_chain/", hits[,"pdb.id"], ".pdb", sep="")
pdbs <- pdbaln(files)
@ 
Inspect the alignment with your favorate alignment viewer (I recomend seaview, available from: )

\paragraph{Unfinished section of draft!}

\paragraph{Atom selection:} 
Bio3d provides an atom selection function (\texttt{atom.select()}) that returns atom and xyz indicies corresponding to the intersection of a hierarchical selection string.  Once an atom selection is made, you can query the properties of the selected atoms, such as their names, residue ids, or coordinates. In a similar fashion, you can set the values of these properties. You can also perform logical operations on atom selections, including finding the intersection or union of two or more atom selections or finding the inverse of the set.

\paragraph{Root mean square deviation (RMSD):}
RMSD is a standard measure of structural distance between coordinate sets.
<<>>=
gaps.xyz <- gap.inspect(pdbs$xyz)
rd.raw <- rmsd(pdbs$xyz[,gaps.xyz$f.inds])
rd.fit <- rmsd(pdbs$xyz[,gaps.xyz$f.inds], fit=TRUE)
range(rd.raw)
range(rd.fit)
@ 

\begin{center}
<<fig=TRUE,echo=TRUE>>=
rmsd.fit <- rmsd(pdbs$xyz[1,gaps.xyz$f.inds], pdbs$xyz[,gaps.xyz$f.inds], fit=TRUE)
hist(rmsd.fit)
@ 
\end{center}


\paragraph{Clustering:}
Cluster with RMSD or see earler section on PCA [reference back]
\begin{center}
<<fig=TRUE,echo=TRUE>>=
plot(hclust(as.dist(rd.fit)), labels=substr(pdbs$id,2,10), ylab="RMSD", xlab="")
@ 
\end{center}



\paragraph{Root mean squared fluctuations (RMSF):}
RMSF is an often used measure of conformational variance.
<<>>=
gaps <- is.na(pdbs$xyz[1,])
rf <- rmsf(xyz[, !gaps])
@ 

\begin{center}
<<fig=TRUE,echo=TRUE>>=
plot.bio3d(rf , sse=sse, ylab="RMSF (A)", xlab="Position")
@ 
\end{center}


\paragraph{Torsion/Dihedral analysis and Difference distance matrix analysis (DDM):}
The conformation of a polypeptide or nucleotide chain can be usefully described in terms of angles of internal rotation around its constituent bonds.
<<>>=
tor <- torsion.pdb(pdb)
## basic Ramachandran plot
@ 
\begin{center}
<<fig=TRUE,echo=TRUE>>=
plot(tor$phi, tor$psi, xlab="phi", ylab="psi")
@ 
\end{center}

<<>>=

a.xyz <- pdbs$xyz[(pdbs$id=="d1i5sa_"),]
b.xyz <- pdbs$xyz[(pdbs$id=="d1i6ia_"),]

gaps.xyz <- is.gap(pdbs$xyz[(pdbs$id=="d1bg2__"),])
gaps.res <- is.gap(pdbs$ali[(pdbs$id=="d1bg2__"),])
resno <- pdbs$resno[(pdbs$id=="d1bg2__"),!gaps.res]

a <- torsion.xyz(a.xyz[!gaps.xyz],atm.inc=1)
b <- torsion.xyz(b.xyz[!gaps.xyz],atm.inc=1)
d.ab <- wrap.tor(a-b)

#par(mfrow=c(3,1))
#plot(resno, d.ab, typ="h")
#plot.bio3d(resno, abs(d.ab), typ="h", sse=sse)

a <- dm(a.xyz[!gaps.xyz])
b <- dm(b.xyz[!gaps.xyz])
ddm <- a - b

#plot(ddm, nlevels=10, grid.col="gray", resnum.1=resno, resnum.2=resno,
#     xlab="1i6i (positions relative to 1bg2)", ylab="1i5s (positions relative to 1bg2)")


@ 

\begin{center}
<<fig=TRUE,echo=TRUE>>=
par(mfrow=c(2,1), mar=c(4,4,0,1))
plot(resno, d.ab, typ="h", xlab="", ylab="Angle")
plot.bio3d(resno, abs(d.ab), typ="h", sse=sse, xlab="Residue", ylab="Angle")
@

<<fig=TRUE,echo=TRUE>>=
plot(ddm, nlevels=10, grid.col="gray", resnum.1=resno, resnum.2=resno,
     xlab="1i6i (positions relative to 1bg2)", ylab="1i5s (positions relative to 1bg2)")
@
 
\end{center}















\subsection{Sequence Conservation Analysis}
The \texttt{read.fasta} and  \texttt{write.fasta} functions can be used to read and write aligned and non aligned sequences in FASTA format.  Whilst the \texttt{aln2html()} function renders a sequence alignment as coloured HTML suitable for viewing with a web browser.

\subsubsection{Sequence Alignment}
The \texttt{seqaln()} function permits the alignment of multiple sequences as obtained from the \texttt{read.fasta()}.  A simple alignment procedure for the sequences in the file \emph{unaligned.fa} would involve the commands:
<<eval=FALSE>>=
aln <-  seqaln(read.fasta("unaligned.fa")) 
@ 

\subsubsection{Residue Conservation Analysis}
To assess the level of sequence conservation at each position in
an alignment, the \emph{similarity}, \emph{identity}, and \emph{entropy} per position can be calculated with the \texttt{conserv()} function.

The \emph{similarity} is defined as the average of the similarity
scores of all pairwise residue comparisons for that position in
the alignment, where the similarity score between any two residues
is the score value between those residues in the chosen
substitution matrix.

The \emph{identity} i.e. the preference for a specific amino acid to be
found at a certain position, is assessed by averaging the identity
scores resulting from all possible pairwise comparisons at that
position in the alignment, where all identical residue comparisons
are given a score of 1 and all other comparisons are given a value
of 0.

\emph{Entropy} is based on Shannon's information entropy. See the
\texttt{entropy} function for further details.

<<>>=
aln <- read.fasta(system.file("examples/kinesin_xray.fa",package="bio3d"))
sim <- conserv(x=aln$ali, method="similarity", sub.matrix="bio3d")
@ 

\begin{center}
<<fig=TRUE,echo=TRUE>>=
sse <- dssp(pdb, resno=FALSE)
plot.bio3d(sim[!is.gap(aln$ali[1,])], sse=sse, xlab="Residue", ylab="Similarity")
@ 
\end{center}



<<eval=FALSE>>=
aln <- read.fasta(system.file("examples/kinesin_xray.fa",package="bio3d"))
write.fasta(seqs=aln$ali[,379:385], file="eg.fa")
aln2html(aln, append=FALSE, file="eg.html")
aln2html(aln, colorscheme="ent", file="eg.html")
@ 


\subsubsection*{Identity, Clustering, Consensus, etc.}

Pairwise identity analysis
dm plot, histogram, ide.filter


Determine the consensus sequence for a given alignment at a given identity cutoff value.

<<>>=
con <- consensus(aln$ali, cutoff = 0.6)
@
Quantifies residue conservation in a given protein sequence
alignment by calculating the degree of amino acid variability in
each column of the alignment.

<<>>=
con.sim <- conserv(x=aln$ali, method="similarity", sub.matrix="bio3d")
##con.ent <- conserv(x=aln$ali, method="entropy10")
@ 
plot conservation 





\subsection{Basic Trajectory Analysis}
Read a CHARMM/X-PLOR/NAMD trajectory file
<<>>=
trtfile <- system.file("examples/hivp.dcd", package="bio3d")
trj <- read.dcd(trtfile)
@ 

Read the starting PDB file to determine atom correspondence
<<>>=
pdbfile <- system.file("examples/hivp.pdb", package="bio3d")
pdb <- read.pdb(pdbfile)
@ 

\subsubsection*{Superposition, RMSD, RMSF, PCA, etc.}
Fit trj on PDB based on residues 23 to 31 and 84 to 87 in both chains
<<>>=
inds <- atom.select(pdb,"///23:31,84:87///CA/")
fit.xyz <- fit.xyz(pdb$xyz, trj, fixed.inds=inds$xyz, mobile.inds=inds$xyz)
@
RMSD of trj frames from PDB
<<>>=
r <- rmsd(a=pdb, b=fit.xyz)
@ 


PCA of trj
<<>>=
pc.trj <- pca.xyz(fit.xyz)
@ 
Overview of PCA results with \texttt{plot.pca}:

Cluster along pc1-2 planes
<<>>=
hc <- hclust(dist(pc.trj$z[,1:2]))
grps <- cutree(hc, k=2) # or h=50
@

\begin{center}
<<fig=TRUE,echo=TRUE>>=
plot(pc.trj, col=c("red","blue")[grps])
@ 
\end{center}


Residue contributions to the first principal component

\begin{center}
<<fig=TRUE,echo=TRUE>>=
plot.bio3d(pc.trj$au[,1], xlab="Residue", ylab="PC1 (A)")
@ 
\end{center}

Write PC trajectory
<<eval=FALSE>>=
a <- mktrj.pca(pc.trj, pc=1, file="pc1_trj.pdb")
@ 


Cross-Correlation Analysis
To be finished.

A more detailed discussion will be the subject of a future vignette.


\subsection*{Session Info}
<<>>=
toLatex(sessionInfo())
@ 

\begin{thebibliography}{9}


\bibitem[Grant \emph{et al.}, 2006]{grant06}
Grant, B.J. and Rodrigues, A.P.D.C and Elsawy, K.M. and Mccammon, A.J. and Caves, L.S.D. (2006)
\textbf{Bio3d: an R package for the comparative analysis of protein structures.}
\emph{Bioinformatics},
\textbf{22}, 2695--2696.


\bibitem[Grant \emph{et al.}, 2007]{grant07}
Grant, B.J. and Mccammon, A.J. and Caves, L.S.D. and Cross, R.A. (2007)
\textbf{Multivariate Analysis of Conserved Sequence-Structure Relationships in Kinesins: Coupling of the Active Site and a Tubulin-binding Sub-domain.}
\emph{J. Mol. Biol.},
\textbf{5}, 1231--1248


\bibitem[Fischer and Karplus, 1992]{cpr}
Fischer, S. and Karplus, M. (1992) 
\textbf{Conjugate peak refinement: an algorithm for finding reaction paths and accurate transition states in systems with many degrees of freedom.}
\emph{Chem. Phys. Lett}, \textbf{194}, 252--261


\bibitem[Hayward and Berendsen, 1989]{dyndom}
Hayward, S. and Berendsen, H. (1998) 
\textbf{Systematic analysis of domain motions in proteins from conformational change: new results on citrate synthase and T4 lysozyme.}
\emph{Proteins}, \textbf{30}, 144--154


\bibitem[Humphrey \emph{et al.}, 1996]{vmd}
Humphrey, W., et al. (1996) 
\textbf{VMD: visual molecular dynamics.}
\emph{J. Mol. Graph}, \textbf{14}, 33--38



\end{thebibliography}




\end{document}

