<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
<title>gap.inspect. bio3d 2.0-1</title>
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<meta name="author" content=" Barry Grant ">

<link href="css/bootstrap.css" rel="stylesheet">
<link href="css/bootstrap-responsive.css" rel="stylesheet">
<link href="css/highlight.css" rel="stylesheet">
<link href="css/staticdocs.css" rel="stylesheet">

<!--[if lt IE 9]>
  <script src="http://html5shim.googlecode.com/svn/trunk/html5.js"></script>
<![endif]-->
  </head>

  <body>
    <div class="navbar">
  <div class="navbar-inner">
    <div class="container">
      <a class="brand" href="http://thegrantlab.org/bio3d/index.html">bio3d 2.0-1</a>
      <div class="nav">
        <ul class="nav">
          <li><a href="index.html"><i class="icon-home icon-white"></i> Index</a></li>
          <li><a href="http://thegrantlab.org/bio3d/download/download.html">Download Bio3D</a></li>
          <li><a href="http://thegrantlab.org">The Grant Lab</a></li>
        </ul>
      </div>
    </div>
  </div>
</div>

    <div class="container">
      <header>
        
      </header>
      
      <h1> Alignment Gap Summary </h1>

<div class="row">
  <div class="span8">
    <h2>Usage</h2>
    <pre>gap.inspect(x)</pre>
    
    <h2>Arguments</h2>
    <dl>
      <dt>x</dt>
      <dd> a matrix or an alignment data structure obtained from
    <code><a href='read.fasta.html'>read.fasta</a></code> or <code><a href='read.fasta.pdb.html'>read.fasta.pdb</a></code>.</dd>
    </dl>
    
    <div class="Description">
      <h2>Description</h2>
      
      <p>Report the number of gaps per sequence and per position for a given alignment.</p>
  
    </div>

    <div class="Details">
      <h2>Details</h2>
      
      <p>Reports the number of gap characters per row (i.e. sequence) and
  per column (i.e. position) for a given <code>alignment</code>. In addition,
  the indices for gap and non-gap containing coloums are returned along
  with a binary matrix indicating the location of gap positions.</p>
  
    </div>

    <div class="Value">
      <h2>Value</h2>
      
      <p>Returns a list object with the following components:
  row a numeric vector detailing the number of gaps per row
    (i.e. sequence).
  col a numeric vector detailing the number of gaps per column
    (i.e. position).
  t.inds  indices for gap containing coloums 
  f.inds  indices for non-gap containing coloums  
  bin a binary numeric matrix with the same dimensions as the
    <code>alignment</code>, with 0 at non-gap positions and 1 at gap
    positions.</p>
  
    </div>

    <div class="References">
      <h2>References</h2>
      
      <p>Grant, B.J. et al. (2006) <em>Bioinformatics</em> <b>22</b>, 2695--2696.</p>
  
    </div>

    <div class="Note">
      <h2>Note</h2>
      
      <p>During alignment, gaps are introduced into sequences that are believed
  to have undergone deletions or insertions with respect to other
  sequences in the alignment.  These gaps, often referred to as
  indels, can be represented with &#145;NA&#146;, a &#145;-&#146; or
  &#145;.&#146; character.</p>
  
      <p>This function gives an overview of gap occurrence and may be useful
  when considering positions or sequences that could/should be excluded
  from further analysis.</p>
  
    </div>
    
    <h2 id="examples">Examples</h2>
    <pre class="examples"><div class='input'>aln <- read.fasta( system.file("examples/hivp_xray.fa",
                   package = "bio3d") )

gap.stats <- gap.inspect(aln$ali)
gap.stats$row # Gaps per sequence
</div>
<div class='output'>  [1] 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 0 0 2 2 2 2 2 2 2 2
 [38] 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 4 3 2 2 4
 [75] 3 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 3 3 2 2 2 2 2 2
[112] 6 6 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 4 4 2 2 2 2 2 2 2
[149] 2 2 2 2 2 6 5 4 4 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
[186] 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 3 2 2 2 2 2 2 2 2 2 2 2 2 2 4 5 2 2 2
[223] 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
[260] 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
[297] 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 3 3 3 2 8 2
[334] 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
[371] 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
[408] 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
</div>
<div class='input'>gap.stats$col # Gaps per position
</div>
<div class='output'>  [1]   5   1   1   1   1   1   0   0   0   0   0   0   0   2   0   0   0   0
 [19]   0   0   2   0   0   1   0   0   0   0   0   0   0   0   0   0   6   1
 [37]   0   0   0   0   0   0   0   0   0   0   0   0  10  10 423 423   0   0
 [55]   0   0   0   0   0   0   0   0   2   0   0   0   0   0   0   0   0   0
 [73]   0   0   0   0   0   0   0   0   0   0   0   0   0   0   2   0   0   0
 [91]   0   0   0   0   0   0   0   0   0   0   0
</div>
<div class='input'>##gap.stats$bin # Binary matrix (1 for gap, 0 for aminoacid)
##aln[,gap.stats$f.inds] # Alignment without gap positions 

plot(gap.stats$col, typ="h", ylab="No. of Gaps") 
</div>
<p><img src='gap.inspect-6.png' alt='' width='400' height='400' /></p></pre>
  </div>
  <div class="span4">
    <!-- <ul>
      <li>gap.inspect</li>
    </ul>
    <ul>
      <li> utilities </li>
    </ul> -->
      
    <h2>See also</h2>
     <code><a href='read.fasta.html'>read.fasta</a></code>, <code><a href='read.fasta.pdb.html'>read.fasta.pdb</a></code>
    
    <h2>Author</h2>
     Barry Grant 
    
  </div>
</div>
      
      <footer>
      <p class='feedback'>What do you think of the documentation?  <a href='http://xinqyao.wufoo.com/forms/documentation-feedback-for-bio3d/def/field1=gap.inspect'>Please let me know by filling out this short online survey</a>.</p>
<p class="pull-right"><a href="#">Back to top</a></p>
<p>Built by <a href="https://github.com/hadley/staticdocs">staticdocs</a>. Styled with <a href="http://twitter.github.com/bootstrap">bootstrap</a>.</p>
      </footer>
    </div>
  </body>
</html>
