<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
<title>ide.filter. bio3d 2.0-1</title>
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<meta name="author" content=" Barry Grant ">

<link href="css/bootstrap.css" rel="stylesheet">
<link href="css/bootstrap-responsive.css" rel="stylesheet">
<link href="css/highlight.css" rel="stylesheet">
<link href="css/staticdocs.css" rel="stylesheet">

<!--[if lt IE 9]>
  <script src="http://html5shim.googlecode.com/svn/trunk/html5.js"></script>
<![endif]-->
  </head>

  <body>
    <div class="navbar">
  <div class="navbar-inner">
    <div class="container">
      <a class="brand" href="http://thegrantlab.org/bio3d/index.html">bio3d 2.0-1</a>
      <div class="nav">
        <ul class="nav">
          <li><a href="index.html"><i class="icon-home icon-white"></i> Index</a></li>
          <li><a href="http://thegrantlab.org/bio3d/download/download.html">Download Bio3D</a></li>
          <li><a href="http://thegrantlab.org">The Grant Lab</a></li>
        </ul>
      </div>
    </div>
  </div>
</div>

    <div class="container">
      <header>
        
      </header>
      
      <h1> Percent Identity Filter </h1>

<div class="row">
  <div class="span8">
    <h2>Usage</h2>
    <pre>ide.filter(aln = NULL, ide = NULL, cutoff = 0.6, verbose = TRUE, ncore = 1, nseg.scale = 1)</pre>
    
    <h2>Arguments</h2>
    <dl>
      <dt>aln</dt>
      <dd> sequence alignment list, obtained from
    <code><a href='seqaln.html'>seqaln</a></code> or <code><a href='read.fasta.html'>read.fasta</a></code>, or an alignment
    character matrix. Not used if &#145;ide&#146; is given.</dd>
      <dt>ide</dt>
      <dd> an optional identity matrix obtained from
    <code><a href='seqidentity.html'>seqidentity</a></code>. </dd>
      <dt>cutoff</dt>
      <dd> a numeric identity cutoff value ranging between 0 and 1. </dd>
      <dt>verbose</dt>
      <dd> logical, if TRUE print details of the clustering process. </dd>
      <dt>ncore</dt>
      <dd> number of CPU cores used to do the calculation.
    <code>ncore>1</code> requires multicore package installed. </dd>
      <dt>nseg.scale </dt>
      <dd> split input data into specified number of segments
    prior to running multiple core calculation. See <code><a href='fit.xyz.html'>fit.xyz</a></code>. </dd>
    </dl>
    
    <div class="Description">
      <h2>Description</h2>
      
      <p>Identify and filter subsets of sequences at a given sequence identity cutoff.</p>
  
    </div>

    <div class="Details">
      <h2>Details</h2>
      
      <p>This function performs hierarchical cluster analysis of a given sequence
  identity matrix &#145;ide&#146;, or the identity matrix calculated from a given
  alignment &#145;aln&#146;, to identify sequences that fall below a given
  identity cutoff value &#145;cutoff&#146;.</p>
  
    </div>

    <div class="Value">
      <h2>Value</h2>
      
      <p>Returns a list object with components:
  indindices of the sequences below the cutoff value.
  treean object of class <code>"hclust"</code>, which describes the
    tree produced by the clustering process. 
  idea numeric matrix with all pairwise identity values.</p>
  
    </div>

    <div class="References">
      <h2>References</h2>
      
      <p>Grant, B.J. et al. (2006) <em>Bioinformatics</em> <b>22</b>, 2695--2696.</p>
  
    </div>
    
    <h2 id="examples">Examples</h2>
    <pre class="examples"><div class='input'>data(kinesin)
attach(kinesin, warn.conflicts=FALSE)

ide.mat <- seqidentity(pdbs)

# Histogram of pairwise identity values
op <- par(no.readonly=TRUE)
par(mfrow=c(2,1))
hist(ide.mat[upper.tri(ide.mat)], breaks=30,xlim=c(0,1),
     main="Sequence Identity", xlab="Identity")

k <- ide.filter(ide=ide.mat, cutoff=0.6)
</div>
<div class='output'>ide.filter(): N clusters @ cutoff =  10 
</div>
<div class='input'>ide.cut <- seqidentity(pdbs$ali[k$ind,])
hist(ide.cut[upper.tri(ide.cut)], breaks=10, xlim=c(0,1),
     main="Sequence Identity", xlab="Identity")
</div>
<p><img src='ide.filter-4.png' alt='' width='400' height='400' /></p>
<div class='input'>
#plot(k$tree, axes = FALSE, ylab="Sequence Identity")
#print(k$ind) # selected
par(op)
detach(kinesin)
</div></pre>
  </div>
  <div class="span4">
    <!-- <ul>
      <li>ide.filter</li>
    </ul>
    <ul>
      <li> utilities </li>
    </ul> -->
      
    <h2>See also</h2>
     <code><a href='read.fasta.html'>read.fasta</a></code>, <code><a href='seqaln.html'>seqaln</a></code>,
  <code><a href='seqidentity.html'>seqidentity</a></code>, <code><a href='entropy.html'>entropy</a></code>, <code><a href='consensus.html'>consensus</a></code> 
    
    <h2>Author</h2>
     Barry Grant 
    
  </div>
</div>
      
      <footer>
      <p class='feedback'>What do you think of the documentation?  <a href='http://xinqyao.wufoo.com/forms/documentation-feedback-for-bio3d/def/field1=ide.filter'>Please let me know by filling out this short online survey</a>.</p>
<p class="pull-right"><a href="#">Back to top</a></p>
<p>Built by <a href="https://github.com/hadley/staticdocs">staticdocs</a>. Styled with <a href="http://twitter.github.com/bootstrap">bootstrap</a>.</p>
      </footer>
    </div>
  </body>
</html>
