<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
<title>pca.xyz. bio3d 1.2</title>
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<meta name="author" content=" Barry Grant ">

<link href="css/bootstrap.css" rel="stylesheet">
<link href="css/bootstrap-responsive.css" rel="stylesheet">
<link href="css/highlight.css" rel="stylesheet">
<link href="css/staticdocs.css" rel="stylesheet">

<!--[if lt IE 9]>
  <script src="http://html5shim.googlecode.com/svn/trunk/html5.js"></script>
<![endif]-->
  </head>

  <body>
    <div class="navbar">
  <div class="navbar-inner">
    <div class="container">
      <a class="brand" href="http://thegrantlab.org/bio3d/index.html">bio3d 1.2</a>
      <div class="nav">
        <ul class="nav">
          <li><a href="index.html"><i class="icon-home icon-white"></i> Index</a></li>
          <li><a href="http://thegrantlab.org/bio3d/download/download.html">Download Bio3D</a></li>
          <li><a href="http://thegrantlab.org">The Grant Lab</a></li>
        </ul>
      </div>
    </div>
  </div>
</div>

    <div class="container">
      <header>
        
      </header>
      
      <h1> Principal Component Analysis </h1>

<div class="row">
  <div class="span8">
    <h2>Usage</h2>
    <pre>pca.xyz(xyz, subset = rep(TRUE, nrow(as.matrix(xyz))))</pre>
    
    <h2>Arguments</h2>
    <dl>
      <dt>xyz</dt>
      <dd> numeric matrix of Cartesian coordinates with a row per
    structure. </dd>
      <dt>subset</dt>
      <dd> an optional vector of numeric indices that selects a
    subset of rows (e.g. experimental structures vs molecular dynamics
    trajectory structures) from the full <code>xyz</code> matrix. Note: the
    full <code>xyz</code> is projected onto this subspace.</dd>
    </dl>
    
    <div class="Description">
      <h2>Description</h2>
      
      <p>Performs principal components analysis (PCA) on a <code>xyz</code>
   numeric data matrix.</p>
  
    </div>

    <div class="Value">
      <h2>Value</h2>
      
      <p>Returns a list with the following components:
  L eigenvalues.
  U eigenvectors (i.e. the x, y, and z variable loadings).
  z scores of the supplied <code>xyz</code> on the pcs.
  au atom-wise loadings (i.e. xyz normalised eigenvectors).
  sdev the standard deviations of the pcs.
  mean the means that were subtracted.</p>
  
    </div>

    <div class="References">
      <h2>References</h2>
      
      <p>Grant, B.J. et al. (2006) <em>Bioinformatics</em> <b>22</b>, 2695--2696.</p>
  
    </div>
    
    <h2 id="examples">Examples</h2>
    <pre class="examples"><div class='input'>
## <strong>Not run</strong>: 
# #-- Read kinesin alignment and structures
# aln <- read.fasta(system.file("examples/kif1a.fa",package="bio3d"))
# pdbs <- read.fasta.pdb(aln)
# 
# # Find core
# core <- core.find(pdbs,
#                   #write.pdbs = TRUE,
#                   verbose=TRUE)
# 
# # Fit structures onto sub 0.5A^3 core
# xyz <- fit.xyz( fixed = pdbs$xyz[1,],
#                 mobile = pdbs,
#                 fixed.inds  = core$c0.5A.xyz,
#                 mobile.inds = core$c0.5A.xyz) 
# ## <strong>End(Not run)</strong>

#-- OR Read previously saved kinesin data
data(kinesin)
attach(kinesin, warn.conflicts=FALSE)

# Remove outlier structures from kinesin alignment
cut.seqs <- which(pdbs$id %in% c("d1n6mb_","d1ry6a_"))

# Ignore gap containing positions
gaps.res <- gap.inspect(pdbs$ali[-cut.seqs,])
gaps.pos <- gap.inspect(pdbs$xyz[-cut.seqs,])


#-- Do PCA
pc.xray <- pca.xyz(xyz[-cut.seqs, gaps.pos$f.inds])

# Plot results (conformer plots & scree plot)
plot(pc.xray)
</div>
<p><img src='pca.xyz-3.png' alt='' width='400' height='400' /></p>
<div class='input'>
## Plot atom wise loadings
plot.bio3d(pc.xray$au[,1], ylab="PC1 (A)")
</div>
<p><img src='pca.xyz-5.png' alt='' width='400' height='400' /></p>
<div class='input'>
#plot.bio3d(gaps.res$f.ind, pc.xray$au[,1],
#           xlab="Alignment Position", ylab="PC1 (A)")

## Plot loadings in relation to reference structure "d1bg2__"
res.ref <- which(!is.gap(pdbs$ali["d1bg2__",]))
res.ind <- which(res.ref %in% gaps.res$f.ind)
par(mfrow = c(3, 1), cex = 0.6, mar = c(3, 4, 1, 1))
plot.bio3d(res.ind, pc.xray$au[,1], sse=sse, ylab="PC1 (A)")
plot.bio3d(res.ind, pc.xray$au[,2], sse=sse, ylab="PC2 (A)")
plot.bio3d(res.ind, pc.xray$au[,3], sse=sse, ylab="PC3 (A)")
</div>
<p><img src='pca.xyz-7.png' alt='' width='400' height='400' /></p>
<div class='input'>


## <strong>Not run</strong>: 
# # Write PC trajectory
# a <- mktrj.pca(pc.xray, pc=1, file="pc1.pdb",
#                resno = pdbs$resno[1, gaps.res$f.inds],
#                resid = aa123(pdbs$ali[1, gaps.res$f.inds]) )
# 
# b <- mktrj.pca(pc.xray, pc=2, file="pc2.pdb",
#                resno = pdbs$resno[1, gaps.res$f.inds],
#                resid = aa123(pdbs$ali[1, gaps.res$f.inds]) )
# 
# c <- mktrj.pca(pc.xray, pc=3, file="pc3.pdb",
#                resno = pdbs$resno[1, gaps.res$f.inds],
#                resid = aa123(pdbs$ali[1, gaps.res$f.inds]) )
# ## <strong>End(Not run)</strong>


</div></pre>
  </div>
  <div class="span4">
    <!-- <ul>
      <li>pca.xyz</li>
    </ul>
    <ul>
      <li> utilities </li><li> multivariate </li>
    </ul> -->
      
    <h2>See also</h2>
     <code><a href='plot.pca.html'>plot.pca</a></code>, <code><a href='mktrj.pca.html'>mktrj.pca</a></code>,
  <code><a href='pca.tor.html'>pca.tor</a></code>, <code><a href='pca.project.html'>pca.project</a></code> 
    
    <h2>Author</h2>
     Barry Grant 
    
  </div>
</div>
      
      <footer>
      <p class='feedback'>What do you think of the documentation?  <a href='http://xinqyao.wufoo.com/forms/documentation-feedback-for-bio3d/def/field1=pca.xyz'>Please let me know by filling out this short online survey</a>.</p>
<p class="pull-right"><a href="#">Back to top</a></p>
<p>Built by <a href="https://github.com/hadley/staticdocs">staticdocs</a>. Styled with <a href="http://twitter.github.com/bootstrap">bootstrap</a>.</p>
      </footer>
    </div>
  </body>
</html>
