<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
<title>pdb2aln.ind. bio3d 2.0</title>
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<meta name="author" content=" Xin-Qiu Yao &amp; Barry Grant ">

<link href="css/bootstrap.css" rel="stylesheet">
<link href="css/bootstrap-responsive.css" rel="stylesheet">
<link href="css/highlight.css" rel="stylesheet">
<link href="css/staticdocs.css" rel="stylesheet">

<!--[if lt IE 9]>
  <script src="http://html5shim.googlecode.com/svn/trunk/html5.js"></script>
<![endif]-->
  </head>

  <body>
    <div class="navbar">
  <div class="navbar-inner">
    <div class="container">
      <a class="brand" href="http://thegrantlab.org/bio3d/index.html">bio3d 2.0</a>
      <div class="nav">
        <ul class="nav">
          <li><a href="index.html"><i class="icon-home icon-white"></i> Index</a></li>
          <li><a href="http://thegrantlab.org/bio3d/download/download.html">Download Bio3D</a></li>
          <li><a href="http://thegrantlab.org">The Grant Lab</a></li>
        </ul>
      </div>
    </div>
  </div>
</div>

    <div class="container">
      <header>
        
      </header>
      
      <h1> Mapping between PDB atomic indices and alignment positions </h1>

<div class="row">
  <div class="span8">
    <h2>Usage</h2>
    <pre>pdb2aln.ind(aln, pdb, inds, ...)</pre>
    
    <h2>Arguments</h2>
    <dl>
      <dt>aln</dt>
      <dd> an alignment list object with <code>id</code> and <code>ali</code>
    components, similar to that generated by <code><a href='read.fasta.html'>read.fasta</a></code>,
    <code><a href='read.fasta.pdb.html'>read.fasta.pdb</a></code>, and <code><a href='seqaln.html'>seqaln</a></code>. </dd>
      <dt>pdb</dt>
      <dd> the PDB object. </dd>
      <dt>inds</dt>
      <dd> a numeric vector with a subset of the alignment position indices. </dd>
      <dt>...</dt>
      <dd> additional arguments for the function <code><a href='pdb2aln.html'>pdb2aln</a></code>. </dd>
    </dl>
    
    <div class="Description">
      <h2>Description</h2>
      
      <p>Find the best alignment between a PDB structure and an 
  existing alignment. Then, given a set of residue indices
  defined for the original alignment, return the equivalent 
  CA atom indices in the PDB coordinates.</p>
  
    </div>

    <div class="Details">
      <h2>Details</h2>
      
      <p>Call the function <code><a href='pdb2aln.html'>pdb2aln</a></code> to align the PDB sequence to
  the existing alignment. Then find the equivalent CA atomic indices in 
  PDB to <code>inds</code>.</p>
  
    </div>

    <div class="Value">
      <h2>Value</h2>
      
      <p>Returns a numeric vector with the equivalent CA atomic indices.</p>
  
    </div>

    <div class="References">
      <h2>References</h2>
      
      <p>Grant, B.J. et al. (2006) <em>Bioinformatics</em> <b>22</b>, 2695--2696.</p>
  
    </div>
    
    <h2 id="examples">Examples</h2>
    <pre class="examples"><div class='input'>##--- Read aligned PDB coordinates (CA only)
aln  <- read.fasta(system.file("examples/kif1a.fa",package="bio3d"))
pdbs <- read.fasta.pdb(aln)
</div>
<div class='output'>pdb/seq: 1   name: http://www.rcsb.org/pdb/files/1bg2.pdb 
pdb/seq: 2   name: http://www.rcsb.org/pdb/files/1i6i.pdb 
   PDB has ALT records, taking A only, rm.alt=TRUE
pdb/seq: 3   name: http://www.rcsb.org/pdb/files/1i5s.pdb 
   PDB has ALT records, taking A only, rm.alt=TRUE
pdb/seq: 4   name: http://www.rcsb.org/pdb/files/2ncd.pdb 
</div>
<div class='input'>
##--- Read PDB coordinate for a new structure (all atoms)
id <- get.pdb("2kin", URLonly=TRUE)
pdb <- read.pdb(id)
</div>
<div class='output'>  HEADER    MOTOR PROTEIN                           28-APR-97   2KIN               
   PDB has ALT records, taking A only, rm.alt=TRUE
</div>
<div class='input'>
# map the non-gap positions
gap.inds <- gap.inspect(pdbs$resno)
npc.inds <- pdb2aln.ind(aln=pdbs, pdb=pdb, id=id, inds=gap.inds$f.inds)
</div>
<div class='output'>muscle -profile -in1 /tmp/RtmpyefGEQ/filec1773c11f253 -in2 /tmp/RtmpyefGEQ/filec177712a22fa -out /tmp/RtmpyefGEQ/filec177728a7e6f</div>
<strong class='warning'>Warning message:
Gaps are found in equivalent positions in PDB</strong>
<div class='input'>
# If gaps are found in PDB sequence with the predefined indices,
# redefine the non-gap positions
ngap.f.inds <- gap.inds$f.inds[!is.na(npc.inds)]
npc.inds <- npc.inds[!is.na(npc.inds)]

##--- fit the atomic coordinates to the aligned X-ray structure
xyz <- fit.xyz(pdbs$xyz[1,], pdb$xyz, atom2xyz(ngap.f.inds), atom2xyz(npc.inds))

## seq2aln(pdbseq(pdb), aln, id = id)
## do we get the same result


</div></pre>
  </div>
  <div class="span4">
    <!-- <ul>
      <li>pdb2aln.ind</li>
    </ul>
    <ul>
      <li> utilities </li>
    </ul> -->
      
    <h2>See also</h2>
     <code><a href='seq2aln.html'>seq2aln</a></code>, <code><a href='seqaln.pair.html'>seqaln.pair</a></code>,
       <code><a href='pdb2aln.html'>pdb2aln</a></code> 
    
    <h2>Author</h2>
     Xin-Qiu Yao & Barry Grant 
    
  </div>
</div>
      
      <footer>
      <p class='feedback'>What do you think of the documentation?  <a href='http://xinqyao.wufoo.com/forms/documentation-feedback-for-bio3d/def/field1=pdb2aln.ind'>Please let me know by filling out this short online survey</a>.</p>
<p class="pull-right"><a href="#">Back to top</a></p>
<p>Built by <a href="https://github.com/hadley/staticdocs">staticdocs</a>. Styled with <a href="http://twitter.github.com/bootstrap">bootstrap</a>.</p>
      </footer>
    </div>
  </body>
</html>
