<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
<title>read.fasta.pdb. bio3d 1.2</title>
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<meta name="author" content=" Barry Grant ">

<link href="css/bootstrap.css" rel="stylesheet">
<link href="css/bootstrap-responsive.css" rel="stylesheet">
<link href="css/highlight.css" rel="stylesheet">
<link href="css/staticdocs.css" rel="stylesheet">

<!--[if lt IE 9]>
  <script src="http://html5shim.googlecode.com/svn/trunk/html5.js"></script>
<![endif]-->
  </head>

  <body>
    <div class="navbar">
  <div class="navbar-inner">
    <div class="container">
      <a class="brand" href="http://thegrantlab.org/bio3d/index.html">bio3d 1.2</a>
      <div class="nav">
        <ul class="nav">
          <li><a href="index.html"><i class="icon-home icon-white"></i> Index</a></li>
          <li><a href="http://thegrantlab.org/bio3d/download/download.html">Download Bio3D</a></li>
          <li><a href="http://thegrantlab.org">The Grant Lab</a></li>
        </ul>
      </div>
    </div>
  </div>
</div>

    <div class="container">
      <header>
        
      </header>
      
      <h1> Read Aligned Structure Data </h1>

<div class="row">
  <div class="span8">
    <h2>Usage</h2>
    <pre>read.fasta.pdb(aln, pdb.path = "", pdbext = "", ncore = 1, nseg.scale = 1, ...)</pre>
    
    <h2>Arguments</h2>
    <dl>
      <dt>aln</dt>
      <dd> an alignment data structure obtained with
    <code><a href='read.fasta.html'>read.fasta</a></code>. </dd>
      <dt>pdb.path</dt>
      <dd> path to PDB files. </dd>
      <dt>pdbext</dt>
      <dd> the file name extention of the PDB files. </dd>
      <dt>ncore </dt>
      <dd> number of CPU cores used to do the calculation.
    <code>ncore>1</code> requires multicore package installed. </dd>
      <dt>nseg.scale </dt>
      <dd> split input data into specified number of segments
    prior to running multiple core calculation. See <code><a href='fit.xyz.html'>fit.xyz</a></code>. </dd>
      <dt>...</dt>
      <dd> other parameters for <code><a href='read.pdb.html'>read.pdb</a></code>. </dd>
    </dl>
    
    <div class="Description">
      <h2>Description</h2>
      
      <p>Read aligned PDB structures and store their C-alpha atom data, including xyz
  coordinates, residue numbers, residue type and B-factors.</p>
  
    </div>

    <div class="Details">
      <h2>Details</h2>
      
      <p>The input <code>aln</code>, produced with <code><a href='read.fasta.html'>read.fasta</a></code>, must
  have identifers (i.e. sequence names) that match the PDB file
  names. For example the sequence corresponding to the structure
  &#147;1bg2.pdb&#148; should have the identifer &#145;1bg2&#146;. See
  examples below.</p>
  
      <p>Sequence miss-matches will generate errors.  Thus, care should be taken
  to ensure that the sequences in the alignment match the sequences in
  their associated PDB files.</p>
  
    </div>

    <div class="Value">
      <h2>Value</h2>
      
      <p>Returns a list of class <code>"3dalign"</code> with the following five
  components:
  xyznumeric matrix of aligned C-alpha coordinates.
  resnocharacter matrix of aligned residue numbers.
  bnumeric matrix of aligned B-factor values.
  chaincharacter matrix of aligned chain identifiers.
  idcharacter vector of PDB sequence/structure names.
  alicharacter matrix of aligned sequences.</p>
  
    </div>

    <div class="References">
      <h2>References</h2>
      
      <p>Grant, B.J. et al. (2006) <em>Bioinformatics</em> <b>22</b>, 2695--2696.</p>
  
    </div>

    <div class="Note">
      <h2>Note</h2>
      
      <p>The sequence character &#145;X&#146; is useful for masking unusual
   or unknown residues, as it can match any other residue type.</p>
  
    </div>
    
    <h2 id="examples">Examples</h2>
    <pre class="examples"><div class='input'># Read sequence alignment
file <- system.file("examples/kif1a.fa",package="bio3d")
aln  <- read.fasta(file)

# Read aligned PDBs
pdbs <- read.fasta.pdb(aln)
</div>
<div class='output'>pdb/seq: 1   name: http://www.rcsb.org/pdb/files/1bg2.pdb 
pdb/seq: 2   name: http://www.rcsb.org/pdb/files/1i6i.pdb 
   PDB has ALT records, taking A only, rm.alt=TRUE
pdb/seq: 3   name: http://www.rcsb.org/pdb/files/1i5s.pdb 
   PDB has ALT records, taking A only, rm.alt=TRUE
pdb/seq: 4   name: http://www.rcsb.org/pdb/files/2ncd.pdb 
</div>
<div class='input'>
# Structure/sequence names/ids
basename( pdbs$id )
</div>
<div class='output'>[1] "1bg2.pdb" "1i6i.pdb" "1i5s.pdb" "2ncd.pdb"
</div>
<div class='input'>
# Alignment positions 335 to 339
pdbs$ali[,335:339]
</div>
<div class='output'>                                       [,1] [,2] [,3] [,4] [,5]
http://www.rcsb.org/pdb/files/1bg2.pdb "L"  "L"  "F"  "G"  "Q" 
http://www.rcsb.org/pdb/files/1i6i.pdb "L"  "R"  "Y"  "A"  "D" 
http://www.rcsb.org/pdb/files/1i5s.pdb "L"  "R"  "Y"  "A"  "D" 
http://www.rcsb.org/pdb/files/2ncd.pdb "L"  "R"  "F"  "A"  "A" 
</div>
<div class='input'>pdbs$resno[,335:339]
</div>
<div class='output'>                                       [,1]  [,2]  [,3]  [,4]  [,5] 
http://www.rcsb.org/pdb/files/1bg2.pdb "316" "317" "318" "319" "320"
http://www.rcsb.org/pdb/files/1i6i.pdb "345" "346" "347" "348" "349"
http://www.rcsb.org/pdb/files/1i5s.pdb "345" "346" "347" "348" "349"
http://www.rcsb.org/pdb/files/2ncd.pdb "661" "662" "663" "664" "665"
</div>
<div class='input'>pdbs$b[,335:339]
</div>
<div class='output'>                                        [,1]  [,2]  [,3]  [,4]  [,5]
http://www.rcsb.org/pdb/files/1bg2.pdb 14.99 15.46  9.61 12.16 26.17
http://www.rcsb.org/pdb/files/1i6i.pdb 19.27 22.65 23.66 20.76 26.02
http://www.rcsb.org/pdb/files/1i5s.pdb 16.08 19.46 16.91 20.48 26.71
http://www.rcsb.org/pdb/files/2ncd.pdb 14.01 22.77 31.48 44.88 54.12
</div>
<div class='input'>
# Alignment C-alpha coordinates for first structure
pdbs$xyz[1,]
</div>
<div class='output'>   [1]  36.940  -0.636  44.620  33.152  -0.588  44.615  31.266   0.469  47.719
  [10]  27.506   0.284  47.954  25.354   2.275  50.357  21.595   1.825  50.574
  [19]  19.617   4.960  51.359  15.989   4.748  52.418  13.920   7.911  52.278
  [28]  10.894   8.753  54.461   7.477   7.518  53.373   5.891  10.279  51.322
  [37]   2.412  11.716  51.731   1.163   9.793  48.697   2.306   6.499  50.214
  [46]   0.802   7.315  53.612  -2.376   8.511  51.867  -2.530   5.223  49.959
  [55]  -2.208   3.346  53.252   1.271   1.904  52.679   3.297   0.553  55.566
  [64]   6.979   0.859  56.414   8.392  -2.592  55.637  12.135  -1.969  56.044
  [73]  13.661  -3.274  59.292  17.222  -2.392  60.288  20.131  -4.137  62.012
  [82]  23.583  -2.724  62.629  24.857   0.088  60.425  24.951  -2.048  57.289
  [91]  21.969  -4.383  57.131  18.355  -3.955  56.054      NA      NA      NA
 [100]      NA      NA      NA      NA      NA      NA      NA      NA      NA
 [109]  15.747  -6.710  56.306  12.973  -7.085  53.743  10.523 -10.008  53.900
 [118]  12.735 -11.689  56.518  15.743 -11.691  54.170      NA      NA      NA
 [127]      NA      NA      NA      NA      NA      NA      NA      NA      NA
 [136]      NA      NA      NA      NA      NA      NA      NA      NA      NA
 [145]  18.828  -9.708  55.288  20.864  -7.574  52.864  24.156  -6.145  54.072
 [154]  25.986  -3.327  52.281  28.836  -0.956  53.091  26.187   1.076  54.913
 [163]  22.439   1.423  55.268  21.006   4.913  55.838  17.668   5.140  57.598
 [172]  14.919   7.608  56.555  16.251  10.670  58.419      NA      NA      NA
 [181]      NA      NA      NA      NA      NA      NA      NA      NA      NA
 [190]      NA      NA      NA      NA      NA      NA      NA      NA      NA
 [199]      NA      NA      NA  19.910  10.495  57.427  21.251  13.911  56.459
 [208]  22.727  15.043  53.171  25.870  15.674  55.233  26.056  12.064  56.466
 [217]  25.563  10.726  52.927  28.308  12.990  51.582  30.817  12.148  54.331
 [226]  30.245   8.411  54.022  29.755   8.195  50.250  31.784  10.933  48.496
 [235]  34.253  12.741  50.777  37.111  10.242  50.274  36.540   9.970  46.515
 [244]  36.902  13.765  46.227  40.240  13.449  48.016  41.320  10.780  45.523
 [253]  40.411  13.011  42.576  42.520  15.800  44.066  45.141  13.097  44.633
 [262]  45.014  12.582  40.863  42.766   9.520  40.571  39.460   8.938  38.776
 [271]  36.188   7.987  40.379  32.494   7.434  39.877  29.455   7.860  42.113
 [280]  26.018   6.597  41.065  22.486   7.057  42.404  20.443   3.943  41.550
 [289]  16.802   3.137  42.223  13.211   3.286  41.082  11.589   6.598  40.181
 [298]  10.892   8.739  43.249  13.318   6.688  45.344  15.586   9.633  46.049
 [307]  18.269   9.671  43.333  18.144  13.392  42.452  17.626  14.417  46.090
 [316]  20.805  12.462  46.825  22.952  13.606  43.879  21.948  17.270  43.912
 [325]  19.154  18.032  46.380  18.650  21.790  46.855  21.621  23.121  44.899
 [334]  21.912  26.316  46.930  20.654  25.301  50.362  23.660  24.686  52.673
 [343]  21.664  22.048  54.578  19.595  20.558  51.768  22.227  20.414  49.030
 [352]  23.160  16.971  47.683  26.444  15.252  46.840  27.350  17.372  43.813
 [361]  27.082  20.798  45.520  29.070  19.503  48.499  31.761  18.078  46.193
 [370]  32.183  21.362  44.291  32.346  23.160  47.615  35.103  20.854  48.871
 [379]  37.132  21.013  45.679  37.352  24.797  45.672  37.877  24.993  49.455
 [388]  40.804  22.615  49.093  42.154  24.732  46.217  41.903  27.880  48.360
 [397]  43.970  26.185  51.079  46.583  25.245  48.486  47.021  28.397  46.372
 [406]  50.796  28.344  46.876  52.765  26.450  44.223  49.835  24.556  42.704
 [415]      NA      NA      NA  48.091  25.795  39.552  44.595  24.339  39.176
 [424]  42.296  24.004  36.171  38.744  22.702  36.534  36.619  21.774  33.521
 [433]  33.088  20.403  33.635  30.826  18.584  31.181  27.136  17.824  31.556
 [442]  25.353  15.373  29.266  22.470  12.910  29.275  21.839   9.447  27.851
 [451]  18.361   8.691  26.550  17.694   5.594  24.407  21.455   4.961  24.391
 [460]  21.952   8.310  22.618  24.092  11.106  24.104  22.940  14.746  24.169
 [469]  25.011  17.703  25.391  23.099  19.524  28.125  25.359  22.559  27.637
 [478]      NA      NA      NA  24.643  22.696  23.887  21.330  21.309  22.632
 [487]  22.601  21.170  19.027  25.082  18.436  20.045  22.701  15.490  19.814
 [496]      NA      NA      NA  23.439  11.761  19.387  27.134  12.212  20.132
 [505]  29.596   9.326  19.878  32.236   7.899  22.205  35.830   7.796  21.019
 [514]  39.166   6.818  22.509  42.096   9.221  22.226  45.777   8.673  21.427
 [523]  46.246   7.240  24.924  43.113   5.034  25.109  41.246   7.555  27.300
 [532]  37.596   7.682  26.147  35.677  10.902  25.438  32.383  12.119  24.008
 [541]  33.210  13.553  20.564  32.262  17.232  20.470  30.949  17.421  24.027
 [550]  31.204  20.854  25.606  33.972  21.190  28.179  33.605  24.326  30.302
 [559]  36.333  25.842  32.490  34.920  26.806  35.868  36.598  28.989  38.485
 [568]  34.157  29.028  41.384  31.341  27.060  43.040  28.851  29.582  41.688
 [577]  30.057  29.031  38.113  29.550  25.287  38.515  26.056  25.635  39.985
 [586]  25.164  28.012  37.147  26.400  25.330  34.735  24.240  22.685  36.413
 [595]  21.318  25.104  36.345  21.833  25.751  32.635  22.029  22.031  31.899
 [604]  18.805  21.518  33.817  17.066  24.227  31.785  18.537  23.049  28.487
 [613]  17.181  19.562  29.255  13.606  20.721  28.538  14.490  21.422  24.916
 [622]  12.624  19.093  22.550  11.121  17.256  25.529   7.332  17.270  25.813
 [631]   7.109  15.473  29.153   9.994  16.265  31.502   8.724  13.907  34.220
 [640]   8.712  10.977  31.815  12.019  12.097  30.293  13.852  12.597  33.585
 [649]  12.782   9.243  35.006  13.853   7.446  31.800  17.222   9.138  31.186
 [658]  20.670   9.294  32.775  22.326  12.587  33.796  26.109  12.632  33.838
 [667]  28.134  15.453  35.347  31.912  15.375  34.897  34.406  17.304  36.998
 [676]  37.973  17.256  35.711  40.502  18.607  38.210  43.948  19.278  36.796
 [685]  46.849  20.725  38.748  50.531  21.475  38.276  52.984  22.412  41.009
 [694]  55.011  25.198  39.394  58.061  24.146  41.422  57.772  20.347  41.665
 [703]      NA      NA      NA      NA      NA      NA      NA      NA      NA
 [712]      NA      NA      NA  56.391  20.206  38.107  54.503  16.984  38.861
 [721]  51.129  17.268  37.113  47.909  15.592  38.247  44.532  15.105  36.611
 [730]  41.461  13.427  37.993  37.989  12.934  36.587  34.988  12.425  38.820
 [739]  31.797  11.326  37.088  28.602  12.012  39.000  25.719   9.945  37.634
 [748]  21.992  10.206  38.308  20.696   7.081  36.565  17.341   6.390  34.908
 [757]  14.646   4.696  36.980  15.401   1.012  37.499  11.695   0.181  37.499
 [766]  10.822  -1.049  34.007   7.406  -2.376  35.095   4.091  -0.474  34.788
 [775]   4.746   1.229  31.439   1.426   2.904  30.622   0.492   4.108  27.143
 [784]  -0.281   2.674  23.710   0.201  -1.066  23.099  -1.719  -4.266  23.858
 [793]  -4.511  -2.832  26.016  -5.506  -5.903  28.029  -2.359  -7.783  26.986
 [802]   1.013  -6.274  27.860   4.169  -6.955  29.867   6.648  -5.923  27.184
 [811]   9.588  -3.821  28.397  13.237  -4.529  29.091  15.633  -3.341  31.771
 [820]  17.319  -0.123  30.641  20.262  -0.836  28.349  22.719   1.190  30.423
 [829]  21.546  -0.237  33.744  21.924  -3.793  32.425  25.316  -2.859  30.962
 [838]  26.286  -1.399  34.324  25.100  -4.582  36.055  27.321  -6.735  33.852
 [847]  30.269  -4.410  34.382  29.898  -4.453  38.179  29.415  -8.235  38.339
 [856]  32.503  -8.707  36.157  34.574  -6.483  38.444  33.158  -8.195  41.535
 [865]  34.261 -11.555  40.046  37.810 -10.306  39.642  37.527 -10.845  35.885
 [874]  40.368  -9.558  33.732  37.855  -8.742  30.989  35.138  -6.211  31.839
 [883]  32.278  -5.631  29.342  32.441  -1.804  29.090  31.392  -1.886  25.435
 [892]  27.916  -3.095  26.448  26.784   0.545  26.420  27.907   4.080  25.599
 [901]  27.709   4.914  29.305  30.020   2.139  30.453  32.375   2.924  27.563
 [910]  32.689   6.608  28.545  33.095   5.664  32.223  35.714   3.035  31.403
 [919]  38.686   4.997  32.763  36.846   5.888  35.986  34.955   2.686  36.822
 [928]  36.813  -0.331  38.108  40.558  -0.784  37.877  42.322   1.737  40.090
 [937]  39.265   3.931  40.596  36.988   4.287  43.595  33.297   3.819  42.957
 [946]  30.340   4.398  45.217  26.872   3.322  44.260  24.093   4.741  46.356
 [955]  21.015   2.610  45.810  17.942   4.547  46.971  14.699   2.925  48.090
 [964]  11.186   3.772  49.179  10.374   1.975  52.461   6.835   1.396  51.144
 [973]   5.017  -1.919  51.234   3.806  -1.012  47.743   7.344  -0.889  46.339
 [982]   8.469  -4.085  48.052   9.291  -5.995  44.883  11.104  -3.025  43.317
 [991]  13.134  -2.495  46.475  14.145  -6.197  46.377  15.412  -5.641  42.815
[1000]  17.382  -2.565  43.846  18.880  -4.358  46.825  19.801  -7.246  44.546
[1009]  21.449  -4.913  42.020  23.509  -3.354  44.785  24.545  -6.830  45.885
[1018]  26.270  -7.439  42.535  29.338  -5.942  44.202  30.022  -8.697  46.749
[1027]  32.358  -7.093  49.297  34.614  -8.456  52.042
</div>
<div class='input'></div></pre>
  </div>
  <div class="span4">
    <!-- <ul>
      <li>read.fasta.pdb</li>
    </ul>
    <ul>
      <li> IO </li>
    </ul> -->
      
    <h2>See also</h2>
    
  <code><a href='read.fasta.html'>read.fasta</a></code>, <code><a href='read.pdb.html'>read.pdb</a></code>,
  <code><a href='core.find.html'>core.find</a></code>, <code><a href='fit.xyz.html'>fit.xyz</a></code>, <code><a href='read.all.html'>read.all</a></code> 
    
    <h2>Author</h2>
     Barry Grant 
    
  </div>
</div>
      
      <footer>
      <p class='feedback'>What do you think of the documentation?  <a href='http://xinqyao.wufoo.com/forms/documentation-feedback-for-bio3d/def/field1=read.fasta.pdb'>Please let me know by filling out this short online survey</a>.</p>
<p class="pull-right"><a href="#">Back to top</a></p>
<p>Built by <a href="https://github.com/hadley/staticdocs">staticdocs</a>. Styled with <a href="http://twitter.github.com/bootstrap">bootstrap</a>.</p>
      </footer>
    </div>
  </body>
</html>
