<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
<title>read.fasta.pdb. bio3d 2.0</title>
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<meta name="author" content=" Barry Grant ">

<link href="css/bootstrap.css" rel="stylesheet">
<link href="css/bootstrap-responsive.css" rel="stylesheet">
<link href="css/highlight.css" rel="stylesheet">
<link href="css/staticdocs.css" rel="stylesheet">

<!--[if lt IE 9]>
  <script src="http://html5shim.googlecode.com/svn/trunk/html5.js"></script>
<![endif]-->
  </head>

  <body>
    <div class="navbar">
  <div class="navbar-inner">
    <div class="container">
      <a class="brand" href="http://thegrantlab.org/bio3d/index.html">bio3d 2.0</a>
      <div class="nav">
        <ul class="nav">
          <li><a href="index.html"><i class="icon-home icon-white"></i> Index</a></li>
          <li><a href="http://thegrantlab.org/bio3d/download/download.html">Download Bio3D</a></li>
          <li><a href="http://thegrantlab.org">The Grant Lab</a></li>
        </ul>
      </div>
    </div>
  </div>
</div>

    <div class="container">
      <header>
        
      </header>
      
      <h1> Read Aligned Structure Data </h1>

<div class="row">
  <div class="span8">
    <h2>Usage</h2>
    <pre>read.fasta.pdb(aln, prefix = "", pdbext = "", ncore = 1, nseg.scale = 1, ...)</pre>
    
    <h2>Arguments</h2>
    <dl>
      <dt>aln</dt>
      <dd> an alignment data structure obtained with
    <code><a href='read.fasta.html'>read.fasta</a></code>. </dd>
      <dt>prefix</dt>
      <dd> prefix to aln$id to locate PDB files. </dd>
      <dt>pdbext</dt>
      <dd> the file name extention of the PDB files. </dd>
      <dt>ncore </dt>
      <dd> number of CPU cores used to do the calculation.
    <code>ncore>1</code> requires multicore package installed. </dd>
      <dt>nseg.scale </dt>
      <dd> split input data into specified number of segments
    prior to running multiple core calculation. See <code><a href='fit.xyz.html'>fit.xyz</a></code>. </dd>
      <dt>...</dt>
      <dd> other parameters for <code><a href='read.pdb.html'>read.pdb</a></code>. </dd>
    </dl>
    
    <div class="Description">
      <h2>Description</h2>
      
      <p>Read aligned PDB structures and store their C-alpha atom data, including xyz
  coordinates, residue numbers, residue type and B-factors.</p>
  
    </div>

    <div class="Details">
      <h2>Details</h2>
      
      <p>The input <code>aln</code>, produced with <code><a href='read.fasta.html'>read.fasta</a></code>, must
  have identifers (i.e. sequence names) that match the PDB file
  names. For example the sequence corresponding to the structure
  &#147;1bg2.pdb&#148; should have the identifer &#145;1bg2&#146;. See
  examples below.</p>
  
      <p>Sequence miss-matches will generate errors.  Thus, care should be taken
  to ensure that the sequences in the alignment match the sequences in
  their associated PDB files.</p>
  
    </div>

    <div class="Value">
      <h2>Value</h2>
      
      <p>Returns a list of class <code>"3dalign"</code> with the following five
  components:
  xyznumeric matrix of aligned C-alpha coordinates.
  resnocharacter matrix of aligned residue numbers.
  bnumeric matrix of aligned B-factor values.
  chaincharacter matrix of aligned chain identifiers.
  idcharacter vector of PDB sequence/structure names.
  alicharacter matrix of aligned sequences.
  residcharacter matrix of aligned 3-letter residue names.</p>
  
    </div>

    <div class="References">
      <h2>References</h2>
      
      <p>Grant, B.J. et al. (2006) <em>Bioinformatics</em> <b>22</b>, 2695--2696.</p>
  
    </div>

    <div class="Note">
      <h2>Note</h2>
      
      <p>The sequence character &#145;X&#146; is useful for masking unusual
   or unknown residues, as it can match any other residue type.</p>
  
    </div>
    
    <h2 id="examples">Examples</h2>
    <pre class="examples"><div class='input'># Read sequence alignment
file <- system.file("examples/kif1a.fa",package="bio3d")
aln  <- read.fasta(file)

# Read aligned PDBs
pdbs <- read.fasta.pdb(aln)
</div>
<div class='output'>pdb/seq: 1   name: http://www.rcsb.org/pdb/files/1bg2.pdb 
pdb/seq: 2   name: http://www.rcsb.org/pdb/files/1i6i.pdb 
   PDB has ALT records, taking A only, rm.alt=TRUE
pdb/seq: 3   name: http://www.rcsb.org/pdb/files/1i5s.pdb 
   PDB has ALT records, taking A only, rm.alt=TRUE
pdb/seq: 4   name: http://www.rcsb.org/pdb/files/2ncd.pdb 
</div>
<div class='input'>
# Structure/sequence names/ids
basename( pdbs$id )
</div>
<div class='output'>[1] "1bg2.pdb" "1i6i.pdb" "1i5s.pdb" "2ncd.pdb"
</div>
<div class='input'>
# Alignment positions 335 to 339
pdbs$ali[,335:339]
</div>
<div class='output'>                                       [,1] [,2] [,3] [,4] [,5]
http://www.rcsb.org/pdb/files/1bg2.pdb "L"  "L"  "F"  "G"  "Q" 
http://www.rcsb.org/pdb/files/1i6i.pdb "L"  "R"  "Y"  "A"  "D" 
http://www.rcsb.org/pdb/files/1i5s.pdb "L"  "R"  "Y"  "A"  "D" 
http://www.rcsb.org/pdb/files/2ncd.pdb "L"  "R"  "F"  "A"  "A" 
</div>
<div class='input'>pdbs$resid[,335:339]
</div>
<div class='output'>                                       [,1]  [,2]  [,3]  [,4]  [,5] 
http://www.rcsb.org/pdb/files/1bg2.pdb "LEU" "LEU" "PHE" "GLY" "GLN"
http://www.rcsb.org/pdb/files/1i6i.pdb "LEU" "ARG" "TYR" "ALA" "ASP"
http://www.rcsb.org/pdb/files/1i5s.pdb "LEU" "ARG" "TYR" "ALA" "ASP"
http://www.rcsb.org/pdb/files/2ncd.pdb "LEU" "ARG" "PHE" "ALA" "ALA"
</div>
<div class='input'>pdbs$resno[,335:339]
</div>
<div class='output'>                                       [,1]  [,2]  [,3]  [,4]  [,5] 
http://www.rcsb.org/pdb/files/1bg2.pdb "316" "317" "318" "319" "320"
http://www.rcsb.org/pdb/files/1i6i.pdb "345" "346" "347" "348" "349"
http://www.rcsb.org/pdb/files/1i5s.pdb "345" "346" "347" "348" "349"
http://www.rcsb.org/pdb/files/2ncd.pdb "661" "662" "663" "664" "665"
</div>
<div class='input'>pdbs$b[,335:339]
</div>
<div class='output'>                                        [,1]  [,2]  [,3]  [,4]  [,5]
http://www.rcsb.org/pdb/files/1bg2.pdb 14.99 15.46  9.61 12.16 26.17
http://www.rcsb.org/pdb/files/1i6i.pdb 19.27 22.65 23.66 20.76 26.02
http://www.rcsb.org/pdb/files/1i5s.pdb 16.08 19.46 16.91 20.48 26.71
http://www.rcsb.org/pdb/files/2ncd.pdb 14.01 22.77 31.48 44.88 54.12
</div>
<div class='input'>
# Alignment C-alpha coordinates for these positions
pdbs$xyz[, atom2xyz(335:339)]
</div>
<div class='output'>                                         [,1]    [,2]   [,3]   [,4]    [,5]
http://www.rcsb.org/pdb/files/1bg2.pdb 18.880  -4.358 46.825 19.801  -7.246
http://www.rcsb.org/pdb/files/1i6i.pdb  4.206  34.360  4.539  3.439  38.005
http://www.rcsb.org/pdb/files/1i5s.pdb 16.417 -33.136  4.337 17.452 -36.750
http://www.rcsb.org/pdb/files/2ncd.pdb  0.185  37.134 46.308 -0.375  34.180
                                         [,6]   [,7]    [,8]   [,9]  [,10]
http://www.rcsb.org/pdb/files/1bg2.pdb 44.546 21.449  -4.913 42.020 23.509
http://www.rcsb.org/pdb/files/1i6i.pdb  5.403  1.775  37.124  8.670  4.660
http://www.rcsb.org/pdb/files/1i5s.pdb  5.131 19.143 -35.570  8.292 15.967
http://www.rcsb.org/pdb/files/2ncd.pdb 48.642 -2.386  36.408 50.962 -4.710
                                         [,11]  [,12]  [,13]   [,14]  [,15]
http://www.rcsb.org/pdb/files/1bg2.pdb  -3.354 44.785 24.545  -6.830 45.885
http://www.rcsb.org/pdb/files/1i6i.pdb  34.826  9.548  7.057  37.666  8.763
http://www.rcsb.org/pdb/files/1i5s.pdb -33.755  9.354 13.829 -36.723  8.388
http://www.rcsb.org/pdb/files/2ncd.pdb  36.941 48.048 -4.498  33.173 47.625
</div>
<div class='input'>
# See 'fit.xyz()' function for actual coordinate superposition
#  e.g. fit to first structure
# xyz <- fit.xyz(pdbs$xyz[1,], pdbs)
# xyz[, atom2xyz(335:339)]
</div></pre>
  </div>
  <div class="span4">
    <!-- <ul>
      <li>read.fasta.pdb</li>
    </ul>
    <ul>
      <li> IO </li>
    </ul> -->
      
    <h2>See also</h2>
    
  <code><a href='read.fasta.html'>read.fasta</a></code>, <code><a href='read.pdb.html'>read.pdb</a></code>,
  <code><a href='core.find.html'>core.find</a></code>, <code><a href='fit.xyz.html'>fit.xyz</a></code>, <code><a href='read.all.html'>read.all</a></code> 
    
    <h2>Author</h2>
     Barry Grant 
    
  </div>
</div>
      
      <footer>
      <p class='feedback'>What do you think of the documentation?  <a href='http://xinqyao.wufoo.com/forms/documentation-feedback-for-bio3d/def/field1=read.fasta.pdb'>Please let me know by filling out this short online survey</a>.</p>
<p class="pull-right"><a href="#">Back to top</a></p>
<p>Built by <a href="https://github.com/hadley/staticdocs">staticdocs</a>. Styled with <a href="http://twitter.github.com/bootstrap">bootstrap</a>.</p>
      </footer>
    </div>
  </body>
</html>
