<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
<title>rmsd. bio3d 1.2</title>
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<meta name="author" content=" Barry Grant ">

<link href="css/bootstrap.css" rel="stylesheet">
<link href="css/bootstrap-responsive.css" rel="stylesheet">
<link href="css/highlight.css" rel="stylesheet">
<link href="css/staticdocs.css" rel="stylesheet">

<!--[if lt IE 9]>
  <script src="http://html5shim.googlecode.com/svn/trunk/html5.js"></script>
<![endif]-->
  </head>

  <body>
    <div class="navbar">
  <div class="navbar-inner">
    <div class="container">
      <a class="brand" href="http://thegrantlab.org/bio3d/index.html">bio3d 1.2</a>
      <div class="nav">
        <ul class="nav">
          <li><a href="index.html"><i class="icon-home icon-white"></i> Index</a></li>
          <li><a href="http://thegrantlab.org/bio3d/download/download.html">Download Bio3D</a></li>
          <li><a href="http://thegrantlab.org">The Grant Lab</a></li>
        </ul>
      </div>
    </div>
  </div>
</div>

    <div class="container">
      <header>
        
      </header>
      
      <h1> Root Mean Square Deviation </h1>

<div class="row">
  <div class="span8">
    <h2>Usage</h2>
    <pre>rmsd(a, b = NULL, a.inds = NULL, b.inds = NULL, fit = FALSE, ncore = 1, nseg.scale = 1)</pre>
    
    <h2>Arguments</h2>
    <dl>
      <dt>a</dt>
      <dd> a numeric vector containing the reference coordinate set for
    comparison with the coordinates in <code>b</code>.  Alternatively, if
    <code>b=NULL</code> then <code>a</code> can be a matrix or list object
    containing multiple coordinates for pairwise comparison. </dd>
      <dt>b</dt>
      <dd> a numeric vector, matrix or list object with an <code>xyz</code>
    component, containing one or more coordinate sets to be compared with
    <code>a</code>. </dd>
      <dt>a.inds </dt>
      <dd> a vector of indices that selects the elements of
    <code>a</code> upon which the calculation should be based. </dd>
      <dt>b.inds </dt>
      <dd> a vector of indices that selects the elements of
    <code>b</code> upon which the calculation should be based. </dd>
      <dt>fit </dt>
      <dd>logical, if TRUE coordinate superposition is performed
    prior to RMSD calculation. </dd>
      <dt>ncore </dt>
      <dd> number of CPU cores used to do the calculation.
    <code>ncore>1</code> requires multicore package installed. </dd>
      <dt>nseg.scale </dt>
      <dd> split input data into specified number of segments
    prior to running multiple core calculation. See <code><a href='fit.xyz.html'>fit.xyz</a></code>. </dd>
    </dl>
    
    <div class="Description">
      <h2>Description</h2>
      
      <p>Calculate the RMSD between coordinate sets.</p>
  
    </div>

    <div class="Details">
      <h2>Details</h2>
      
      <p>RMSD is a standard measure of structural distance between
  coordinate sets.</p>
  
      <p>Structure <code>a[a.inds]</code> and <code>b[b.inds]</code> should have the
  same length.</p>
  
      <p>A least-squares fit is performed prior to RMSD calculation by setting
  <code>fit=TRUE</code>.  See the function <code>fit.xyz</code> for more
  details of the fitting process.</p>
  
    </div>

    <div class="Value">
      <h2>Value</h2>
      
      <p>Returns a numeric vector of RMSD value(s).</p>
  
    </div>

    <div class="References">
      <h2>References</h2>
      
      <p>Grant, B.J. et al. (2006) <em>Bioinformatics</em> <b>22</b>, 2695--2696.</p>
  
    </div>
    
    <h2 id="examples">Examples</h2>
    <pre class="examples"><div class='input'># -- Calculate RMSD between two or more structures
aln <- read.fasta(system.file("examples/kif1a.fa",package="bio3d"))
pdbs <- read.fasta.pdb(aln)
</div>
<div class='output'>pdb/seq: 1   name: http://www.rcsb.org/pdb/files/1bg2.pdb 
pdb/seq: 2   name: http://www.rcsb.org/pdb/files/1i6i.pdb 
   PDB has ALT records, taking A only, rm.alt=TRUE
pdb/seq: 3   name: http://www.rcsb.org/pdb/files/1i5s.pdb 
   PDB has ALT records, taking A only, rm.alt=TRUE
pdb/seq: 4   name: http://www.rcsb.org/pdb/files/2ncd.pdb 
</div>
<div class='input'>
# Indices 
gaps <- unique( which(is.na(pdbs$xyz),arr.ind=TRUE)[,2] )
inds <- c(1:ncol(pdbs$xyz))[-gaps]

# RMSD between structure 1 and structure 2
rmsd(a=pdbs$xyz[1,], b=pdbs$xyz[2,], a.inds=inds, b.inds=inds)
</div>
<div class='output'>[1] 48.901
</div>
<div class='input'>rmsd(a=pdbs$xyz[1,], b=pdbs$xyz[2,], a.inds=inds, b.inds=inds, fit=TRUE)
</div>
<div class='output'>[1] 2.654
</div>
<div class='input'>
# RMSD between structure 1 and structures 2 and 3
rmsd(a=pdbs$xyz[1,], b=pdbs$xyz[2:3,], a.inds=inds, b.inds=inds)
</div>
<div class='output'>http://www.rcsb.org/pdb/files/1i6i.pdb http://www.rcsb.org/pdb/files/1i5s.pdb 
                                48.901                                 48.533 
</div>
<div class='input'>
# RMSD between structure 1 and all structures in alignment
rmsd(a=pdbs$xyz[1,], b=pdbs, a.inds=inds, b.inds=inds, fit=TRUE)
</div>
<div class='output'>http://www.rcsb.org/pdb/files/1bg2.pdb http://www.rcsb.org/pdb/files/1i6i.pdb 
                                 0.000                                  2.654 
http://www.rcsb.org/pdb/files/1i5s.pdb http://www.rcsb.org/pdb/files/2ncd.pdb 
                                 2.412                                  2.775 
</div>
<div class='input'>
# pairwise RMSD
rmsd(pdbs$xyz[,inds])
</div>
<div class='output'>       [,1]   [,2]   [,3]   [,4]
[1,]  0.000 48.901 48.533 63.343
[2,] 48.901  0.000 58.171 53.186
[3,] 48.533 58.171  0.000 88.198
[4,] 63.343 53.186 88.198  0.000
</div></pre>
  </div>
  <div class="span4">
    <!-- <ul>
      <li>rmsd</li>
    </ul>
    <ul>
      <li> utilities </li>
    </ul> -->
      
    <h2>See also</h2>
     <code><a href='fit.xyz.html'>fit.xyz</a></code>, <code><a href='fit.xyz.html'>rot.lsq</a></code>,
  <code><a href='read.pdb.html'>read.pdb</a></code>, <code><a href='read.fasta.pdb.html'>read.fasta.pdb</a></code> 
    
    <h2>Author</h2>
     Barry Grant 
    
  </div>
</div>
      
      <footer>
      <p class='feedback'>What do you think of the documentation?  <a href='http://xinqyao.wufoo.com/forms/documentation-feedback-for-bio3d/def/field1=rmsd'>Please let me know by filling out this short online survey</a>.</p>
<p class="pull-right"><a href="#">Back to top</a></p>
<p>Built by <a href="https://github.com/hadley/staticdocs">staticdocs</a>. Styled with <a href="http://twitter.github.com/bootstrap">bootstrap</a>.</p>
      </footer>
    </div>
  </body>
</html>
