<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
<title>seqaln. bio3d 2.0-1</title>
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<meta name="author" content=" Barry Grant ">

<link href="css/bootstrap.css" rel="stylesheet">
<link href="css/bootstrap-responsive.css" rel="stylesheet">
<link href="css/highlight.css" rel="stylesheet">
<link href="css/staticdocs.css" rel="stylesheet">
<link href="css/theme.css" rel="stylesheet">

<!--[if lt IE 9]>
  <script src="http://html5shim.googlecode.com/svn/trunk/html5.js"></script>
<![endif]-->
  </head>

  <body>
    <div class="container">
<div class="navbar navbar-default" role="navigation">
  <div class="navbar-header">
      <a class="navbar-brand" href="http://thegrantlab.org/bio3d">bio3d 2.0-1</a>
  </div>
  <div class="navbar-collapse collapse">
    <ul class="nav navbar-nav">
        <li class="active"><a href="index.html"> Index</a></li>
        <li><a href="http://thegrantlab.org/bio3d/download">Download Bio3D</a></li>
        <li><a href="http://thegrantlab.org">The Grant Lab</a></li>
    </ul>
  </div>
</div>
</div>

    <div class="container">
      <header>
        
      </header>
      
      <h1> Sequence Alignment with MUSCLE</h1>

<div class="row">
  <div class="span8">
    <h2>Usage</h2>
    <pre>seqaln(aln, id = NULL, exefile = "muscle", outfile = "aln.fa", protein = TRUE, seqgroup = FALSE, 
  refine = FALSE, extra.args = "", verbose = FALSE)</pre>
    
    <h2>Arguments</h2>
    <dl>
      <dt>aln</dt>
      <dd> a sequence character matrix, as obtained from
    <code><a href='seqbind.html'>seqbind</a></code>, or an alignment list object as obtained from
      <code><a href='read.fasta.html'>read.fasta</a></code>. </dd>
      <dt>id</dt>
      <dd> a vector of sequence names to serve as sequence identifers. </dd>
      <dt>exefile</dt>
      <dd> file path to the &#145;MUSCLE&#146; program on your system (i.e.
    how is &#145;MUSCLE&#146; invoked). </dd>
      <dt>outfile</dt>
      <dd> name of &#145;FASTA&#146; output file to which alignment
    should be written. </dd>
      <dt>protein</dt>
      <dd> logical, if TRUE the input sequences are assumed to be
    protein not DNA or RNA. </dd>
      <dt>seqgroup</dt>
      <dd> logical, if TRUE similar sequences are grouped
    together in the output. </dd>
      <dt>refine</dt>
      <dd> logical, if TRUE the input sequences are assumed to already
  be aligned, and only tree dependent refinement is performed. </dd>
      <dt>extra.args</dt>
      <dd> a single character string containing extra command
    line arguments for the alignment program. </dd>
      <dt>verbose</dt>
      <dd> logical, if TRUE &#145;MUSCLE&#146; warning and error
    messages are printed. </dd>
    </dl>
    
    <div class="Description">
      <h2>Description</h2>
      
      <p>Create multiple alignments of amino acid or nucleotide sequences
  according to the method of Edgar.</p>
  
    </div>

    <div class="Details">
      <h2>Details</h2>
      
      <p>Sequence alignment attempts to arrange the sequences of protein,
  DNA or RNA, to highlight regions of shared similarity that may
  reflect functional, structural, and/or evolutionary relationships
  between the sequences.</p>
  
      <p>Aligned sequences are represented as rows within a matrix. Gaps
  (&#145;-&#146;) are inserted between the aminoacids or nucleotides so that
  equivalent characters are positioned in the same column.</p>
  
      <p>This function calls the &#145;MUSCLE&#146; program, to perform a multiple
  sequence alignment, which MUST BE INSTALLED on your system and in the
  search path for executables.</p>
  
      <p>If you have a large number of input sequences (a few thousand), or they are
  very long, the default settings may be too slow for practical
  use. A good compromise between speed and accuracy is to run just the
  first two iterations of the &#145;MUSCLE&#146; algorithm by setting the
  <code>extra.args</code> argument to &#147;-maxiters 2&#148;.</p>
  
      <p>You can set &#145;MUSCLE&#146; to improve an existing alignment by setting
  <code>refine</code> to TRUE.</p>
  
      <p>To inspect the sequence clustering used by &#145;MUSCLE&#146; to produce
  alignments, include &#147;-tree2 tree.out&#148; in the <code>extra.args</code>
  argument. You can then load the &#147;tree.out&#148; file with the
  &#145;read.tree&#146; function from the &#145;ape&#146; package.</p>
  
    </div>

    <div class="Value">
      <h2>Value</h2>
      
      <p>A list with two components:
  ali an alignment character matrix with a row per sequence and
    a column per equivalent aminoacid/nucleotide. 
  ids sequence names as identifers.</p>
  
    </div>

    <div class="References">
      <h2>References</h2>
      
      <p>Grant, B.J. et al. (2006) <em>Bioinformatics</em> <b>22</b>, 2695--2696.</p>
  
      <p>&#145;MUSCLE&#146; is the work of Edgar:
  Edgar (2004) <em>Nuc. Acid. Res.</em> <b>32</b>, 1792--1797.</p>
  
      <p>Full details of the &#145;MUSCLE&#146; algorithm, along with download and
  installation instructions can be obtained from:
  <a href = 'http://www.drive5.com/muscle'>http://www.drive5.com/muscle</a>.</p>
  
    </div>

    <div class="Note">
      <h2>Note</h2>
      
      <p>A system call is made to the &#145;MUSCLE&#146; program, which must be
  installed on your system and in the search path for executables.</p>
  
    </div>
    
    <h2 id="examples">Examples</h2>
    <pre class="examples"><div class='input'>##-- Read a folder/directory of PDB files
#pdb.path <- "my_dir_of_pdbs"
#files  <- list.files(path=pdb.path ,
#                     pattern=".pdb",
#                     full.names=TRUE)

##-- Use online files
files <- get.pdb(c("4q21","1ftn"), URLonly=TRUE)

##-- Extract and store sequences
raw <- NULL
for(i in 1:length(files)) {
  pdb <- read.pdb(files[i])
  raw <- seqbind(raw, pdbseq(pdb) )
}
</div>
<div class='output'>  HEADER    ONCOGENE PROTEIN                        25-SEP-91   4Q21               
  HEADER    PROTO-ONCOGENE                          13-MAR-97   1FTN               
</div>
<div class='input'>
##-- Align these sequences
aln <- seqaln(raw, id=files, outfile="seqaln.fa")

##-- Read Aligned PDBs storing coordinate data
pdbs <- read.fasta.pdb(aln) 
</div>
<div class='output'>pdb/seq: 1   name: http://www.rcsb.org/pdb/files/4q21.pdb 
pdb/seq: 2   name: http://www.rcsb.org/pdb/files/1ftn.pdb 
</div>
<div class='input'>
## Sequence identity
seqidentity(aln)
</div>
<div class='output'>      [,1]  [,2]
[1,] 1.000 0.352
[2,] 0.352 1.000
</div>
<div class='input'>
## Do further analysis...
## Note that all the above can be done with the pdbaln() function.
## pdbs <- pdbaln( c("first.pdb","second.pdb", "third.pdb") )



##-- For identical sequences with masking use a custom matrix
aln <- list(ali=seqbind(c("X","C","X","X","A","G","K"),
                        c("C","-","A","X","G","X","X","K")),
            id=c("a","b"))

temp <- seqaln(aln=aln, outfile="temp.fas", protein=TRUE,
               extra.args= paste("-matrix",
                system.file("matrices/custom.mat", package="bio3d"),
                "-gapopen -3.0 ",
                "-gapextend -0.5",
                "-center 0.0") )

</div></pre>
  </div>
  <div class="span3">
    <!-- <ul>
      <li>seqaln</li>
    </ul>
    <ul>
      <li> utilities </li>
    </ul> -->
      
    <h2>See also</h2>
     <code><a href='read.fasta.html'>read.fasta</a></code>, <code><a href='read.fasta.pdb.html'>read.fasta.pdb</a></code>,
  <code><a href='seqbind.html'>seqbind</a></code>

    
    <h2>Author</h2>
     Barry Grant 
    
  </div>
</div>
      
      <footer>
      <p class='feedback'>What do you think of the documentation?  <a href='http://thegrantlab.wufoo.com/forms/documentation-feedback-for-bio3d/def/field1=seqaln'>Please let me know by filling out this short online survey</a>.</p>
<p class="pull-right"><a href="#">Back to top</a></p>
<p>Built by <a href="https://github.com/hadley/staticdocs">staticdocs</a>. Styled with <a href="http://twitter.github.com/bootstrap">bootstrap</a>.</p>
      </footer>
    </div>
  </body>
</html>
