<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
<title>seqaln.pair. bio3d 2.0</title>
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<meta name="author" content=" Barry Grant ">

<link href="css/bootstrap.css" rel="stylesheet">
<link href="css/bootstrap-responsive.css" rel="stylesheet">
<link href="css/highlight.css" rel="stylesheet">
<link href="css/staticdocs.css" rel="stylesheet">

<!--[if lt IE 9]>
  <script src="http://html5shim.googlecode.com/svn/trunk/html5.js"></script>
<![endif]-->
  </head>

  <body>
    <div class="navbar">
  <div class="navbar-inner">
    <div class="container">
      <a class="brand" href="http://thegrantlab.org/bio3d/index.html">bio3d 2.0</a>
      <div class="nav">
        <ul class="nav">
          <li><a href="index.html"><i class="icon-home icon-white"></i> Index</a></li>
          <li><a href="http://thegrantlab.org/bio3d/download/download.html">Download Bio3D</a></li>
          <li><a href="http://thegrantlab.org">The Grant Lab</a></li>
        </ul>
      </div>
    </div>
  </div>
</div>

    <div class="container">
      <header>
        
      </header>
      
      <h1> Sequence Alignment of Identical Protein Sequences </h1>

<div class="row">
  <div class="span8">
    <h2>Usage</h2>
    <pre>seqaln.pair(aln, extra.args = "", ...)</pre>
    
    <h2>Arguments</h2>
    <dl>
      <dt>aln</dt>
      <dd> a sequence character matrix, as obtained from
    <code><a href='seqbind.html'>seqbind</a></code>, or an alignment list object as obtained from
      <code><a href='read.fasta.html'>read.fasta</a></code>. </dd>
      <dt>extra.args</dt>
      <dd> a single character string containing extra command
    line arguments for the alignment program. </dd>
      <dt>...</dt>
      <dd> additional arguments for the function <code><a href='seqaln.html'>seqaln</a></code>. </dd>
    </dl>
    
    <div class="Description">
      <h2>Description</h2>
      
      <p>Create multiple alignments of amino acid sequences according to the
  method of Edgar.</p>
  
    </div>

    <div class="Details">
      <h2>Details</h2>
      
      <p>This function is intended for the alignment of identical sequences only.
  For standard alignment see the related function <code><a href='seqaln.html'>seqaln</a></code>.</p>
  
      <p>This function is useful for determining the equivalences between
  sequences and structures. For example in aligning a PDB sequence to an
  existing multiple sequence alignment, where one would first mask the
  alignment sequences and then run the alignment to determine equivalences.</p>
  
    </div>

    <div class="Value">
      <h2>Value</h2>
      
      <p>A list with two components:
  ali an alignment character matrix with a row per sequence and
    a column per equivalent aminoacid/nucleotide. 
  ids sequence names as identifers.</p>
  
    </div>

    <div class="References">
      <h2>References</h2>
      
      <p>Grant, B.J. et al. (2006) <em>Bioinformatics</em> <b>22</b>, 2695--2696.</p>
  
      <p>&#145;MUSCLE&#146; is the work of Edgar:
  Edgar (2004) <em>Nuc. Acid. Res.</em> <b>32</b>, 1792--1797.</p>
  
      <p>Full details of the &#145;MUSCLE&#146; algorithm, along with download and
  installation instructions can be obtained from:
  <a href = 'http://www.drive5.com/muscle'>http://www.drive5.com/muscle</a>.</p>
  
    </div>

    <div class="Note">
      <h2>Note</h2>
      
      <p>A system call is made to the &#145;MUSCLE&#146; program, which must be
  installed on your system and in the search path for executables.</p>
  
    </div>
    
    <h2 id="examples">Examples</h2>
    <pre class="examples"><div class='input'>##- Aligning a PDB sequence to an existing sequence alignment


##- Simple example
aln <- list(ali=seqbind(c("X","C","X","X","A","G","K"),
                        c("C","-","A","X","G","X","X","K")),
            id=c("a","b"))

seqaln.pair(aln)
</div>
<div class='output'>muscle -in /tmp/RtmpyefGEQ/filec1771061163e -out aln.fa -matrix /u2/xqyao/Project/bio3d/ver_devel/bio3d/inst/matrices/custom.mat -gapopen -3.0  -gapextend -0.5 -center 0.0  -seqtype protein</div>
<div class='output'>$id
[1] "a" "b"

$ali
  [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10]
a "X"  "C"  "X"  "X"  "A"  "-"  "G"  "-"  "-"  "K"  
b "-"  "C"  "-"  "-"  "A"  "X"  "G"  "X"  "X"  "K"  

attr(,"class")
[1] "fasta"
</div>
<div class='input'></div></pre>
  </div>
  <div class="span4">
    <!-- <ul>
      <li>seqaln.pair</li>
    </ul>
    <ul>
      <li> utilities </li>
    </ul> -->
      
    <h2>See also</h2>
     <code><a href='seqaln.html'>seqaln</a></code>, <code><a href='read.fasta.html'>read.fasta</a></code>,
  <code><a href='read.fasta.pdb.html'>read.fasta.pdb</a></code>, <code><a href='seqbind.html'>seqbind</a></code>

    
    <h2>Author</h2>
     Barry Grant 
    
  </div>
</div>
      
      <footer>
      <p class='feedback'>What do you think of the documentation?  <a href='http://xinqyao.wufoo.com/forms/documentation-feedback-for-bio3d/def/field1=seqaln.pair'>Please let me know by filling out this short online survey</a>.</p>
<p class="pull-right"><a href="#">Back to top</a></p>
<p>Built by <a href="https://github.com/hadley/staticdocs">staticdocs</a>. Styled with <a href="http://twitter.github.com/bootstrap">bootstrap</a>.</p>
      </footer>
    </div>
  </body>
</html>
